diff --git a/libgloss/xstormy16/crt0.s b/libgloss/xstormy16/crt0.s
index f8adb98cd..a9623d71e 100644
--- a/libgloss/xstormy16/crt0.s
+++ b/libgloss/xstormy16/crt0.s
@@ -1,6 +1,7 @@
 # XSTORMY16 startup code
 
 # Interrupt vectors at 0x8000.
+	.section .note.GNU-stack,"",@progbits
 	.section .int_vec,"ax"
 	.global	_start
 	.align 1
diff --git a/libgloss/xstormy16/crt0_stub.s b/libgloss/xstormy16/crt0_stub.s
index 14ec47146..388f3a880 100644
--- a/libgloss/xstormy16/crt0_stub.s
+++ b/libgloss/xstormy16/crt0_stub.s
@@ -1,5 +1,7 @@
 # XSTORMY16 startup code for GDB stub.
 
+	.section .note.GNU-stack,"",@progbits
+
 # CPU Data for Sanyo EVA debugger at 0x7F00
         .section .cpudata,"ax"
         .byte   0x00,0x02,0x80,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00
diff --git a/libgloss/xstormy16/crti.s b/libgloss/xstormy16/crti.s
index 3713ba3a1..bbc7e4d6b 100644
--- a/libgloss/xstormy16/crti.s
+++ b/libgloss/xstormy16/crti.s
@@ -1,3 +1,4 @@
+	.section .note.GNU-stack,"",@progbits
 	.section	.init,"ax"
 	.align 1
 	.globl _init
diff --git a/libgloss/xstormy16/crtn.s b/libgloss/xstormy16/crtn.s
index 22d71b8df..10663d058 100644
--- a/libgloss/xstormy16/crtn.s
+++ b/libgloss/xstormy16/crtn.s
@@ -1,3 +1,4 @@
+	.section .note.GNU-stack,"",@progbits
 	.section	.init,"ax"
         ret
 
diff --git a/libgloss/xstormy16/sim_malloc_start.s b/libgloss/xstormy16/sim_malloc_start.s
index 143053f3b..676eb3336 100644
--- a/libgloss/xstormy16/sim_malloc_start.s
+++ b/libgloss/xstormy16/sim_malloc_start.s
@@ -1,3 +1,5 @@
+	.section .note.GNU-stack,"",@progbits
+
 # This file just defines __malloc_start for newlib for the simulator.
 # The simulator has RAM up to the I/O area at 0x7F00.
 	.globl __malloc_start
diff --git a/libgloss/xstormy16/syscalls.S b/libgloss/xstormy16/syscalls.S
index 8f4175f31..d428ff53d 100644
--- a/libgloss/xstormy16/syscalls.S
+++ b/libgloss/xstormy16/syscalls.S
@@ -2,6 +2,7 @@
 
 #include <syscall.h>
 
+	.section .note.GNU-stack,"",@progbits
 	.text
 
 	.globl _exit
diff --git a/newlib/libc/machine/xstormy16/setjmp.S b/newlib/libc/machine/xstormy16/setjmp.S
index f2f7210cd..be0f9b627 100644
--- a/newlib/libc/machine/xstormy16/setjmp.S
+++ b/newlib/libc/machine/xstormy16/setjmp.S
@@ -6,6 +6,7 @@
 ;; 
 	.file "setjmp.S"
 
+	.section .note.GNU-stack,"",@progbits
 	.section .text
 	.align 1
 	.global setjmp
