diff --git a/libgloss/mcore/kill.c b/libgloss/mcore/kill.c
index 833fe89d7..848668eaa 100644
--- a/libgloss/mcore/kill.c
+++ b/libgloss/mcore/kill.c
@@ -14,6 +14,8 @@
  */
 #include "glue.h"
 
+void    _exit (int __status) _ATTRIBUTE ((__noreturn__));
+
 /*
  * kill -- go out via exit...
  */
diff --git a/libgloss/mcore/putnum.c b/libgloss/mcore/putnum.c
index 10e298b95..a1ef99063 100644
--- a/libgloss/mcore/putnum.c
+++ b/libgloss/mcore/putnum.c
@@ -13,6 +13,8 @@
  * they apply.
  */
 #include "glue.h"
+extern _print (char *ptr);
+
 
 /*
  * putnum -- print a 32 bit number in hex
diff --git a/libgloss/mcore/raise.c b/libgloss/mcore/raise.c
index e39207792..39df08414 100644
--- a/libgloss/mcore/raise.c
+++ b/libgloss/mcore/raise.c
@@ -13,6 +13,8 @@
  * they apply.
  */
 #include "glue.h"
+extern int _kill (int, int);
+extern int _getpid (void);
 
 int
 _raise (int sig)
diff --git a/libgloss/mcore/sbrk.c b/libgloss/mcore/sbrk.c
index 844958984..27cb1e74d 100644
--- a/libgloss/mcore/sbrk.c
+++ b/libgloss/mcore/sbrk.c
@@ -16,6 +16,9 @@
 #include <sys/stat.h>
 #include "glue.h"
 
+extern int _write (int, char *, int);
+
+
 caddr_t
 _sbrk (size_t incr)
 {
