diff --git a/libgloss/m32r/m32r-stub.c b/libgloss/m32r/m32r-stub.c
index acc89ed6a..0e2188720 100644
--- a/libgloss/m32r/m32r-stub.c
+++ b/libgloss/m32r/m32r-stub.c
@@ -96,9 +96,9 @@
  *
  * external low-level support routines
  */
-extern void putDebugChar ();	/* write a single character      */
+extern void putDebugChar (unsigned char);	/* write a single character      */
 extern int getDebugChar ();	/* read and return a single char */
-extern void exceptionHandler ();	/* assign an exception handler   */
+extern void exceptionHandler (int, int[16]);	/* assign an exception handler   */
 
 /*****************************************************************************
  * BUFMAX defines the maximum number of characters in inbound/outbound buffers
diff --git a/libgloss/cris/lcrt0.c b/libgloss/cris/lcrt0.c
index 0aae3c279..b4edcd99d 100644
--- a/libgloss/cris/lcrt0.c
+++ b/libgloss/cris/lcrt0.c
@@ -102,7 +102,7 @@ extern void __init__start (void) __attribute ((weak));
 extern void __aout__ctors (void) __attribute ((weak));
 extern int main (int argc, char **argv, char **env);
 
-static void start1 () __asm__ ("__start1") __attribute ((__used__));
+static void start1 (int, char **, char **) __asm__ ("__start1") __attribute ((__used__));
 static void
 start1 (int argc, char **argv, char **env)
 {
