
diff --git a/newlib/libc/machine/riscv/stpcpy.c b/newlib/libc/machine/riscv/stpcpy.c
index 9243457b25a2..0c545623ba9e 100644
--- a/newlib/libc/machine/riscv/stpcpy.c
+++ b/newlib/libc/machine/riscv/stpcpy.c
@@ -1,5 +1,5 @@
-#include <string.h>
 #include <stdbool.h>
+#include "sys/string.h"
 
 char *stpcpy(char *dst, const char *src)
 {
diff --git a/newlib/libc/machine/riscv/strcpy.c b/newlib/libc/machine/riscv/strcpy.c
index f770493fbc2d..856b66ebc801 100644
--- a/newlib/libc/machine/riscv/strcpy.c
+++ b/newlib/libc/machine/riscv/strcpy.c
@@ -9,8 +9,8 @@
    http://www.opensource.org/licenses.
 */
 
-#include <string.h>
 #include <stdbool.h>
+#include "sys/string.h"
 
 char *strcpy(char *dst, const char *src)
 {
diff --git a/newlib/libc/machine/riscv/sys/asm.h b/newlib/libc/machine/riscv/sys/asm.h
index 0a354b220517..a9792e964ebc 100644
--- a/newlib/libc/machine/riscv/sys/asm.h
+++ b/newlib/libc/machine/riscv/sys/asm.h
@@ -12,7 +12,9 @@
 #ifndef _SYS_ASM_H
 #define _SYS_ASM_H
 
+#if !__ASSEMBLER__
 #include <stdint.h>
+#endif
 
 /*
  * Macros to handle different pointer/register sizes for 32/64-bit code
@@ -22,13 +24,17 @@
 # define SZREG	8
 # define REG_S sd
 # define REG_L ld
+#if !__ASSEMBLER__
 typedef uint64_t uintxlen_t;
+#endif
 #elif __riscv_xlen == 32
 # define PTRLOG 2
 # define SZREG	4
 # define REG_S sw
 # define REG_L lw
+#if !__ASSEMBLER__
 typedef uint32_t uintxlen_t;
+#endif
 #else
 # error __riscv_xlen must equal 32 or 64
 #endif


