diff --git a/libgloss/m32r/m32r-stub.c b/libgloss/m32r/m32r-stub.c
index acc89ed6a..0e2188720 100644
--- a/libgloss/m32r/m32r-stub.c
+++ b/libgloss/m32r/m32r-stub.c
@@ -96,9 +96,9 @@
  *
  * external low-level support routines
  */
-extern void putDebugChar ();	/* write a single character      */
+extern void putDebugChar (unsigned char);	/* write a single character      */
 extern int getDebugChar ();	/* read and return a single char */
-extern void exceptionHandler ();	/* assign an exception handler   */
+extern void exceptionHandler (int, void *);	/* assign an exception handler   */
 
 /*****************************************************************************
  * BUFMAX defines the maximum number of characters in inbound/outbound buffers
diff --git a/libgloss/cris/lcrt0.c b/libgloss/cris/lcrt0.c
index 0aae3c279..b4edcd99d 100644
--- a/libgloss/cris/lcrt0.c
+++ b/libgloss/cris/lcrt0.c
@@ -102,7 +102,7 @@ extern void __init__start (void) __attribute ((weak));
 extern void __aout__ctors (void) __attribute ((weak));
 extern int main (int argc, char **argv, char **env);
 
-static void start1 () __asm__ ("__start1") __attribute ((__used__));
+static void start1 (int, char **, char **) __asm__ ("__start1") __attribute ((__used__));
 static void
 start1 (int argc, char **argv, char **env)
 {
diff --git a/libgloss/mn10300/cygmon.c b/libgloss/mn10300/cygmon.c
index 513115cd6..c164b2728 100644
--- a/libgloss/mn10300/cygmon.c
+++ b/libgloss/mn10300/cygmon.c
@@ -61,8 +61,7 @@ int __trap0(unsigned long func, unsigned long p1, unsigned long p2, unsigned lon
 }
 
 void *
-get_mem_info (mem)
-     struct s_mem *mem;
+get_mem_info (struct s_mem *mem)
 {
   unsigned long totmem = 0, topmem = 0;
   register int numbanks;
diff --git a/libgloss/mn10300/trap.h b/libgloss/mn10300/trap.h
index 27636c004..386dbcb0d 100644
--- a/libgloss/mn10300/trap.h
+++ b/libgloss/mn10300/trap.h
@@ -1,5 +1,5 @@
 #include "syscall.h"
 
-int __trap0 ();
+int __trap0 (int, int, int, int);
 
 #define TRAP0(f, p1, p2, p3) __trap0(f, (int)(p1), (int)(p2), (int)(p3))
diff --git a/libgloss/mn10300/utime.c b/libgloss/mn10300/utime.c
index 3fce7e364..94aa061d2 100644
--- a/libgloss/mn10300/utime.c
+++ b/libgloss/mn10300/utime.c
@@ -5,9 +5,7 @@
 
 
 int
-utime (path, times)
-     const char *path;
-     char *times;
+utime (const char *path, char *times)
 {
   return TRAP0 (SYS_utime, path, times, 0);
 }
diff --git a/libgloss/iq2000/trap.h b/libgloss/iq2000/trap.h
index 27636c004..bf92be69c 100644
--- a/libgloss/iq2000/trap.h
+++ b/libgloss/iq2000/trap.h
@@ -1,5 +1,5 @@
 #include "syscall.h"
 
-int __trap0 ();
+int __trap0 (long, long, long, long);
 
-#define TRAP0(f, p1, p2, p3) __trap0(f, (p1), (p2), (p3))
+#define TRAP0(f, p1, p2, p3) __trap0(f, (long)(p1), (long)(p2), (long)(p3))
