diff --git a/gcc/testsuite/lib/target-supports.exp b/gcc/testsuite/lib/target-supports.exp
index 5026c5906cd..0e346652ed5 100644
--- a/gcc/testsuite/lib/target-supports.exp
+++ b/gcc/testsuite/lib/target-supports.exp
@@ -1094,7 +1094,8 @@ proc check_effective_target_freorder {} {
     } "-freorder-blocks-and-partition"]
     && [check_no_compiler_messages fprofile_use_freorder object {
 	void foo (void) { }
-    } "-fprofile-use -freorder-blocks-and-partition -Wno-missing-profile"] } {
+    } "-fprofile-use -freorder-blocks-and-partition -Wno-missing-profile"] 
+    && ! [istarget fr30*-*-*] && ! [istarget v850*-*-*] } {
 	return 1
     }
     return 0
diff --git a/gcc/testsuite/lib/target-supports.exp b/gcc/testsuite/lib/target-supports.exp
index 04a2a8e8659..8ffbb48635a 100644
--- a/gcc/testsuite/lib/target-supports.exp
+++ b/gcc/testsuite/lib/target-supports.exp
@@ -564,6 +564,7 @@ proc check_effective_target_trampolines { } {
 	 || [istarget nvptx-*-*]
 	 || [istarget hppa2.0w-hp-hpux11.23]
 	 || [istarget hppa64-hp-hpux11.23]
+	 || [istarget ft32-*-*]
 	 || [istarget pru-*-*]
 	 || [istarget bpf-*-*] } {
 	return 0;
