diff --git a/gcc/testsuite/lib/target-supports.exp b/gcc/testsuite/lib/target-supports.exp
index 5026c5906cd..0e346652ed5 100644
--- a/gcc/testsuite/lib/target-supports.exp
+++ b/gcc/testsuite/lib/target-supports.exp
@@ -1094,7 +1094,8 @@ proc check_effective_target_freorder {} {
     } "-freorder-blocks-and-partition"]
     && [check_no_compiler_messages fprofile_use_freorder object {
 	void foo (void) { }
-    } "-fprofile-use -freorder-blocks-and-partition -Wno-missing-profile"] } {
+    } "-fprofile-use -freorder-blocks-and-partition -Wno-missing-profile"] 
+    && ! [istarget fr30*-*-*] && ! [istarget v850*-*-*] } {
 	return 1
     }
     return 0
diff --git a/gcc/testsuite/lib/target-supports.exp b/gcc/testsuite/lib/target-supports.exp
index 236128bf5de..04251018ddb 100644
--- a/gcc/testsuite/lib/target-supports.exp
+++ b/gcc/testsuite/lib/target-supports.exp
@@ -556,6 +556,7 @@ proc check_effective_target_trampolines { } {
     if { [istarget avr-*-*]
 	 || [istarget msp430-*-*]
 	 || [istarget nvptx-*-*]
+	 || [istarget ft32-*-*]
 	 || [istarget pru-*-*]
 	 || [istarget bpf-*-*] } {
 	return 0;
diff --git a/gcc/testsuite/gcc.c-torture/execute/ieee/ieee.exp b/gcc/testsuite/gcc.c-torture/execute/ieee/ieee.exp
index 6d3c371d6ef..8b3684d823e 100644
--- a/gcc/testsuite/gcc.c-torture/execute/ieee/ieee.exp
+++ b/gcc/testsuite/gcc.c-torture/execute/ieee/ieee.exp
@@ -53,6 +53,10 @@ if { [istarget "alpha*-*-*"]
   lappend additional_flags "-mieee"
 }
 
+if [istarget rx-*-*] then {
+  lappend additional_flags "-mnofpu"
+}
+
 if { ![check_effective_target_signal] } {
     lappend additional_flags "-DSIGNAL_SUPPRESS"
 }
