diff --git a/gcc/config/c6x/c6x.cc b/gcc/config/c6x/c6x.cc
index b577317c1e3..32e7abea895 100644
--- a/gcc/config/c6x/c6x.cc
+++ b/gcc/config/c6x/c6x.cc
@@ -1323,7 +1323,7 @@ legitimize_pic_address (rtx orig, rtx reg, rtx picreg)
 
       if (GET_CODE (addr) == CONST_INT)
 	{
-	  gcc_assert (! reload_in_progress && ! reload_completed);
+	  gcc_assert (! ira_in_progress && ! reload_completed);
 	  addr = force_reg (Pmode, addr);
 	}
 
@@ -1349,7 +1349,7 @@ expand_move (rtx *operands, machine_mode mode)
   rtx dest = operands[0];
   rtx op = operands[1];
 
-  if ((reload_in_progress | reload_completed) == 0
+  if ((lra_in_progress | reload_completed) == 0
       && GET_CODE (dest) == MEM && GET_CODE (op) != REG)
     operands[1] = force_reg (mode, op);
   else if (mode == SImode && symbolic_operand (op, SImode))
@@ -1364,7 +1364,7 @@ expand_move (rtx *operands, machine_mode mode)
 	    }
 	  else
 	    {
-	      rtx temp = (reload_completed || reload_in_progress
+	      rtx temp = (reload_completed || lra_in_progress
 			  ? dest : gen_reg_rtx (Pmode));
 
 	      operands[1] = legitimize_pic_address (op, temp,
@@ -6751,9 +6751,6 @@ c6x_regno_reg_class (int reg)
 #undef TARGET_LEGITIMATE_ADDRESS_P
 #define TARGET_LEGITIMATE_ADDRESS_P c6x_legitimate_address_p
 
-#undef TARGET_LRA_P
-#define TARGET_LRA_P hook_bool_void_false
-
 #undef TARGET_IN_SMALL_DATA_P
 #define TARGET_IN_SMALL_DATA_P c6x_in_small_data_p
 #undef	TARGET_ASM_SELECT_RTX_SECTION
