diff --git a/gcc/ira-build.cc b/gcc/ira-build.cc
index ac84f512460..991dfbb3da3 100644
--- a/gcc/ira-build.cc
+++ b/gcc/ira-build.cc
@@ -1851,14 +1851,15 @@ create_insn_allocnos (rtx x, rtx outer, bool output_p)
 	  ira_allocno_t a;
 
 	  if ((a = ira_curr_regno_allocno_map[regno]) == NULL)
+	    a = ira_create_allocno (regno, false, ira_curr_loop_tree_node);
+
+	  /* This used to only trigger at allocno creation which seems
+	     wrong.  We care about the WMODE propery across all the uses.  */
+	  if (outer != NULL && GET_CODE (outer) == SUBREG)
 	    {
-	      a = ira_create_allocno (regno, false, ira_curr_loop_tree_node);
-	      if (outer != NULL && GET_CODE (outer) == SUBREG)
-		{
-		  machine_mode wmode = GET_MODE (outer);
-		  if (partial_subreg_p (ALLOCNO_WMODE (a), wmode))
-		    ALLOCNO_WMODE (a) = wmode;
-		}
+	      machine_mode wmode = GET_MODE (outer);
+	      if (partial_subreg_p (ALLOCNO_WMODE (a), wmode))
+		ALLOCNO_WMODE (a) = wmode;
 	    }
 
 	  ALLOCNO_NREFS (a)++;
