diff --git a/gcc/config/riscv/thead.cc b/gcc/config/riscv/thead.cc
index f43bfdbb354..1f25af96e7f 100644
--- a/gcc/config/riscv/thead.cc
+++ b/gcc/config/riscv/thead.cc
@@ -678,7 +678,7 @@ th_memidx_classify_address_index (struct riscv_address_info *info, rtx x,
     {
       type = (GET_CODE (XEXP (offset, 0)) == SIGN_EXTEND)
 	     ? ADDRESS_REG_REG : ADDRESS_REG_UREG;
-      shift = exact_log2 (INTVAL (XEXP (x, 1)));
+      shift = exact_log2 (INTVAL (offset));
       offset = XEXP (XEXP (x, 0), 0);
     }
   /* (ashift:X (reg:X) (const_int shift)) */
diff --git a/gcc/testsuite/gcc.target/riscv/pr124984.c b/gcc/testsuite/gcc.target/riscv/pr124984.c
new file mode 100644
index 00000000000..7d1124893b7
--- /dev/null
+++ b/gcc/testsuite/gcc.target/riscv/pr124984.c
@@ -0,0 +1,10 @@
+/* { dg-do compile { target bitint } } */
+/* { dg-options "-mcpu=thead-c906 -ftracer -fno-tree-dominator-opts -O2 -fharden-conditional-branches" } */
+
+int g;
+
+void
+foo(int s)
+{
+  g = 0x258d9b711584058072a68c6c9d0ff404fd077b87d59ee75dd2e349825d3a288c61a054e3aa1804c26adb2b9ce2287f17888af3aca18a430bcb4f334672553aecd010e5618a1293691bfd6f3e1d0d8ea3dbdb9176a53d1f657032263707cae4e5e70badc7769309b006495fbd3e76ca25f4d02df66107dc351e928wb >> (s & 1023);
+}
