diff --git a/gcc/ifcvt.cc b/gcc/ifcvt.cc
index 5ea25f8fbe7..722cd8c1bbe 100644
--- a/gcc/ifcvt.cc
+++ b/gcc/ifcvt.cc
@@ -3316,13 +3316,14 @@ noce_try_sign_mask (struct noce_if_info *if_info)
     otherwise NULL_RTX for other RTX_CODE.  */
 
 static rtx
-get_base_reg (rtx exp)
+get_base_reg_or_constant (rtx exp)
 {
   if (REG_P (exp))
     return exp;
   else if (SUBREG_P (exp))
     return SUBREG_REG (exp);
-
+  else if (CONST_INT_P (exp))
+    return exp;
   return NULL_RTX;
 }
 
@@ -3396,24 +3397,27 @@ noce_try_cond_arith (struct noce_if_info *if_info)
   op = GET_CODE (a);
 
   /* Canonicalize x = (z op y) : y to x = (y op z) : y */
-  a_op1 = get_base_reg (XEXP (a, 1));
+  a_op1 = get_base_reg_or_constant (XEXP (a, 1));
   if (a_op1 && rtx_equal_p (a_op1, b) && COMMUTATIVE_ARITH_P (a))
     {
       std::swap (XEXP (a, 0), XEXP (a, 1));
-      a_op1 = get_base_reg (XEXP (a, 1));
+      a_op1 = get_base_reg_or_constant (XEXP (a, 1));
     }
 
   if (a_op1 == NULL_RTX)
     goto fail;
 
   /* Ensure the cond is of form: x = (y op z) : y */
-  a_op0 = get_base_reg (XEXP (a, 0));
+  a_op0 = get_base_reg_or_constant (XEXP (a, 0));
   if (!(a_op0 && rtx_equal_p (a_op0, b)))
     goto fail;
 
   start_sequence ();
 
-  target = gen_reg_rtx (GET_MODE (XEXP (a, op != AND)));
+  if (CONST_INT_P (XEXP (a, 1)))
+    target = gen_reg_rtx (GET_MODE (XEXP (a, 0)));
+  else
+    target = gen_reg_rtx (GET_MODE (XEXP (a, op != AND)));
 
   /* AND requires !cond, instead we swap ops around.  */
   target = noce_emit_cmove (if_info, target, code,
diff --git a/gcc/testsuite/gcc.target/riscv/pr125731.c b/gcc/testsuite/gcc.target/riscv/pr125731.c
new file mode 100644
index 00000000000..19205c2f943
--- /dev/null
+++ b/gcc/testsuite/gcc.target/riscv/pr125731.c
@@ -0,0 +1,47 @@
+/* { dg-do compile } */
+/* { dg-options "-O2 -march=rv64gcbv_zicond -mabi=lp64d" { target rv64} } */
+/* { dg-options "-O2 -march=rv32gcbv_zicond -mabi=ilp32" { target rv32} } */
+
+long fun_not1 (int a, long b)
+{
+    if (!(a & 1))
+	b ^= 8;
+    return b;
+}
+
+long fun_not2 (short a, int b)
+{
+    if (!(a & 1))
+	b ^= 8;
+    return b;
+}
+
+long fun_not21 (long a, long b)
+{
+    if (!(a & 1))
+	b ^= 8;
+    return b;
+}
+
+long fun_not22 (long a, int b)
+{
+    if (!(a & 1))
+	b ^= 8;
+    return b;
+}
+
+long fun_not3 (int a, long b)
+{
+    if (!(a & 1))
+	b |= 8;
+    return b;
+}
+
+long fun_not4 (int a, long b)
+{
+    if (!(a & 1))
+	b &= 8;
+    return b;
+}
+
+/* { dg-final { scan-assembler-not "\tbne" } } */

