diff --git a/gcc/ifcvt.cc b/gcc/ifcvt.cc
index 37e75746193..37f7d4d0daf 100644
--- a/gcc/ifcvt.cc
+++ b/gcc/ifcvt.cc
@@ -3189,12 +3189,14 @@ noce_cond_zero_binary_op_supported (rtx op)
     otherwise NULL_RTX for other RTX_CODE.  */
 
 static rtx
-get_base_reg (rtx exp)
+get_base_reg_or_const_int (rtx exp)
 {
   if (REG_P (exp))
     return exp;
   else if (SUBREG_P (exp))
     return SUBREG_REG (exp);
+  else if (CONST_INT_P (exp))
+    return exp;
 
   return NULL_RTX;
 }
@@ -3232,6 +3234,7 @@ noce_try_cond_arith (struct noce_if_info *if_info)
   rtx_insn *seq;
   rtx_code op;
   machine_mode mode = GET_MODE (if_info->x);
+  machine_mode tmp_mode;
 
   /* Scalar integral modes are only supported here.
      Could support scalar floating point but that
@@ -3269,24 +3272,29 @@ noce_try_cond_arith (struct noce_if_info *if_info)
   op = GET_CODE (a);
 
   /* Canonicalize x = (z op y) : y to x = (y op z) : y */
-  a_op1 = get_base_reg (XEXP (a, 1));
+  a_op1 = get_base_reg_or_const_int (XEXP (a, 1));
   if (a_op1 && rtx_equal_p (a_op1, b) && COMMUTATIVE_ARITH_P (a))
     {
       std::swap (XEXP (a, 0), XEXP (a, 1));
-      a_op1 = get_base_reg (XEXP (a, 1));
+      a_op1 = get_base_reg_or_const_int (XEXP (a, 1));
     }
 
   if (a_op1 == NULL_RTX)
     goto fail;
 
   /* Ensure the cond is of form: x = (y op z) : y */
-  a_op0 = get_base_reg (XEXP (a, 0));
+  a_op0 = get_base_reg_or_const_int (XEXP (a, 0));
   if (!(a_op0 && rtx_equal_p (a_op0, b)))
     goto fail;
 
   start_sequence ();
 
-  target = gen_reg_rtx (GET_MODE (XEXP (a, op != AND)));
+  /* XEXP (a, op != AND) might be a CONST_INT and thus VOIDmode.  In
+     that case fall back to WORD_MODE and hope for the best.  The worry
+     is asserts in the various move paths tripping on a mode mismatch.  */
+  tmp_mode = GET_MODE (XEXP (a, op != AND));
+  tmp_mode = tmp_mode == VOIDmode ? word_mode : tmp_mode;
+  target = gen_reg_rtx (tmp_mode);
 
   /* AND requires !cond, instead we swap ops around.  */
   target = noce_emit_cmove (if_info, target, code,
diff --git a/gcc/testsuite/gcc.target/riscv/pr125731.c b/gcc/testsuite/gcc.target/riscv/pr125731.c
new file mode 100644
index 00000000000..d041f900d50
--- /dev/null
+++ b/gcc/testsuite/gcc.target/riscv/pr125731.c
@@ -0,0 +1,18 @@
+/* { dg-do compile } */
+/* { dg-options "-march=rv64gcb_zicond -mabi=lp64d" { target rv64 } } */
+/* { dg-options "-march=rv32gcb_zicond -mabi=ilp32" { target rv32 } } */
+/* { dg-skip-if "" { *-*-* } { "-O0" "-Og"} } */
+
+long fun_not1 (int a, long b)
+{
+    if (!(a & 1))
+        b ^= 8;
+    return b;
+}
+
+
+/* { dg-final { scan-assembler-times "\\tczero" 1 } } */
+/* { dg-final { scan-assembler-not "\\tbeq\\t" } } */
+/* { dg-final { scan-assembler-not "\\tbne\\t" } } */
+
+
