diff --git a/gcc/match.pd b/gcc/match.pd
index e0a2399ca65..422b8951704 100644
--- a/gcc/match.pd
+++ b/gcc/match.pd
@@ -12231,3 +12231,15 @@ and,
    long as the requested element is within range.  */
 (simplify (IFN_VEC_EXTRACT (vec_duplicate @0) INTEGER_CST@1)
  @0)
+
+(for op (plus bit_and bit_ior bit_xor mult)
+ (simplify
+  (cond (ne @0 integer_zerop) (op @1 @2) @1)
+  (op @1
+    (cond (ne @0 { build_zero_cst (TREE_TYPE (@0)); }) @2 { build_zero_cst (type); }))))
+
+(for op (minus rshift lshift lrotate rrotate)
+ (simplify
+  (cond (ne @0 integer_zerop) (op @1 @2) @1)
+  (op @1
+    (cond (ne @0 { build_zero_cst (TREE_TYPE (@0)); }) @2 { build_zero_cst (type); }))))
diff --git a/gcc/testsuite/gcc.target/riscv/pr123322.c b/gcc/testsuite/gcc.target/riscv/pr123322.c
new file mode 100644
index 00000000000..606d573c1b1
--- /dev/null
+++ b/gcc/testsuite/gcc.target/riscv/pr123322.c
@@ -0,0 +1,18 @@
+/* { dg-do compile } */
+/* { dg-options "-march=rv32gcb_zicond -mabi=ilp32 -mbranch-cost=4" { target { rv32 } } } */
+/* { dg-options "-march=rv64gcb_zicond -mabi=lp64d -mbranch-cost=4" { target { rv64 } } } */
+/* { dg-skip-if "" { *-*-* } { "-O0" "-Og"} } */
+
+#define TEST(NAME,OP) int f##NAME(int a, int b, int c) { if (c) a OP b; return a; }
+
+TEST(and, &=)
+TEST(xor, ^=)
+TEST(ior, |=)
+TEST(rshift, >>=)
+TEST(lshift, <<=)
+TEST(plus, +=)
+TEST(minus, -=)
+TEST(mult, *=)
+
+/* Each test should have precisely one czero.  */ 
+/* { dg-final { scan-assembler-times "czero" 8 } } */
