diff --git a/gcc/config/riscv/riscv.cc b/gcc/config/riscv/riscv.cc
index 3baf0a936b5..1fa311bbcac 100644
--- a/gcc/config/riscv/riscv.cc
+++ b/gcc/config/riscv/riscv.cc
@@ -4281,7 +4281,9 @@ riscv_rtx_costs (rtx x, machine_mode mode, int outer_code, int opno ATTRIBUTE_UN
 		been to duplicate the operation than to CSE the constant.
 	     3. TODO: make cost more accurate specially if riscv_const_insns
 		returns > 1.  */
-	  if (outer_code == SET || GET_MODE (x) == VOIDmode)
+	  if (outer_code == COMPARE)
+	    *total = COSTS_N_INSNS (cost);
+	  else if (outer_code == SET || GET_MODE (x) == VOIDmode)
 	    *total = COSTS_N_INSNS (1);
 	}
       else /* The instruction will be fetched from the constant pool.  */
diff --git a/gcc/testsuite/gcc.target/riscv/pr124029.c b/gcc/testsuite/gcc.target/riscv/pr124029.c
new file mode 100644
index 00000000000..a74258e0c0d
--- /dev/null
+++ b/gcc/testsuite/gcc.target/riscv/pr124029.c
@@ -0,0 +1,18 @@
+/* { dg-do compile */
+/* { dg-options "-march=rv64gcvb_zicond -mabi=lp64d" { target { rv64 } } } */
+/* { dg-options "-march=rv32gcvb_zicond -mabi=ilp32" { target { rv32 } } } */
+/* { dg-skip-if "" { *-*-* } { "-O0" "-O1" "-Og" } } */
+
+void frob (void);
+void
+refine_subpel (int cost, int b_chroma_me)
+{
+
+  int bcost = bcost = (1 << 28);
+  if (b_chroma_me && cost < bcost)
+    frob ();
+}
+
+/* { dg-final { scan-assembler-not "addi\t" } } */
+/* { dg-final { scan-assembler-not "seqz\t" } } */
+
diff --git a/gcc/testsuite/gcc.target/riscv/rvv/autovec/struct/struct_vect-2.c b/gcc/testsuite/gcc.target/riscv/rvv/autovec/struct/struct_vect-2.c
index 52ded08faf5..75586775add 100644
--- a/gcc/testsuite/gcc.target/riscv/rvv/autovec/struct/struct_vect-2.c
+++ b/gcc/testsuite/gcc.target/riscv/rvv/autovec/struct/struct_vect-2.c
@@ -15,8 +15,8 @@
 /* { dg-final { scan-assembler-times {vsseg3e16\.v} 8 } } */
 /* { dg-final { scan-assembler-times {vsseg4e16\.v} 4 } } */
 /* { dg-final { scan-assembler-times {vsseg5e16\.v} 4 } } */
-/* { dg-final { scan-assembler-times {vsseg6e16\.v} 4 } } */
-/* { dg-final { scan-assembler-times {vsseg7e16\.v} 2 } } */
+/* { dg-final { scan-assembler-times {vsseg6e16\.v} 8 } } */
+/* { dg-final { scan-assembler-times {vsseg7e16\.v} 4 } } */
 /* { dg-final { scan-assembler-times {vsseg8e16\.v} 2 } } */
 /* { dg-final { scan-assembler-times {vsetivli\s+zero,\s*8,\s*e16,\s*m1,\s*t[au],\s*m[au]} 14 } } */
 /* { dg-final { scan-assembler-not {vsetvli} } } */
