diff --git a/gcc/ifcvt.cc b/gcc/ifcvt.cc
index 1df5d54efa36..7c2d27fc66a3 100644
--- a/gcc/ifcvt.cc
+++ b/gcc/ifcvt.cc
@@ -1595,6 +1595,30 @@ noce_try_store_flag_constants (struct noce_if_info *if_info)
       b = XEXP (b, 1);
     }
 
+  /* Handle conditional add.  We're narrowly focused right now on cases
+     where the constant is a power of two which enables us to shift left
+     the result of the comparison and unconditionally add that to the
+     common term.  */
+  if (GET_CODE (a) == PLUS
+      && GET_CODE (b) == REG
+      && rtx_equal_p (XEXP (a, 0), b)
+      && CONST_INT_P (XEXP (a, 1)))
+    {
+      common = XEXP (a, 0);
+      a = XEXP (a, 1);
+      b = CONST0_RTX (GET_MODE (common));
+    }
+
+  if (GET_CODE (b) == PLUS
+      && GET_CODE (a) == REG
+      && rtx_equal_p (XEXP (b, 0), a)
+      && CONST_INT_P (XEXP (b, 1)))
+    {
+      common = XEXP (b, 0);
+      b = XEXP (b, 1);
+      a = CONST0_RTX (GET_MODE (common));
+    }
+
   if (!noce_simple_bbs (if_info))
     return false;
 
@@ -1725,6 +1749,25 @@ noce_try_store_flag_constants (struct noce_if_info *if_info)
 					gen_int_mode (ifalse, mode), target,
 					if_info->x, 0, OPTAB_WIDEN);
 	}
+      else if (STORE_FLAG_VALUE == 1
+	       && common
+	       && pow2p_hwi (diff))
+	{
+	  /* Right now TARGET has the value 1/0.  A left shift will produce
+	     DIFF.  */
+	  target = expand_simple_binop (mode, ASHIFT, target,
+				        GEN_INT (exact_log2 (diff)),
+					NULL_RTX, 0, OPTAB_WIDEN);
+
+	  /* Now we just need to add to common part.  If TARGET is NULL,
+	     we'll end the sequence and return failure below.  Similarly
+	     if this expansion step fails.  */
+	  if (target)
+	    target = expand_simple_binop (mode, PLUS,
+					  target, common,
+					  NULL_RTX, 0, OPTAB_WIDEN);
+	}
+ 
       /* Other cases are not beneficial when the original A and B are PLUS
 	 expressions.  */
       else if (common)
@@ -3189,12 +3232,14 @@ noce_cond_zero_binary_op_supported (rtx op)
     otherwise NULL_RTX for other RTX_CODE.  */
 
 static rtx
-get_base_reg (rtx exp)
+get_base_reg_or_const_int (rtx exp)
 {
   if (REG_P (exp))
     return exp;
   else if (SUBREG_P (exp))
     return SUBREG_REG (exp);
+  else if (CONST_INT_P (exp))
+    return exp;
 
   return NULL_RTX;
 }
@@ -3232,6 +3277,7 @@ noce_try_cond_arith (struct noce_if_info *if_info)
   rtx_insn *seq;
   rtx_code op;
   machine_mode mode = GET_MODE (if_info->x);
+  machine_mode tmp_mode;
 
   /* Scalar integral modes are only supported here.
      Could support scalar floating point but that
@@ -3269,24 +3315,29 @@ noce_try_cond_arith (struct noce_if_info *if_info)
   op = GET_CODE (a);
 
   /* Canonicalize x = (z op y) : y to x = (y op z) : y */
-  a_op1 = get_base_reg (XEXP (a, 1));
+  a_op1 = get_base_reg_or_const_int (XEXP (a, 1));
   if (a_op1 && rtx_equal_p (a_op1, b) && COMMUTATIVE_ARITH_P (a))
     {
       std::swap (XEXP (a, 0), XEXP (a, 1));
-      a_op1 = get_base_reg (XEXP (a, 1));
+      a_op1 = get_base_reg_or_const_int (XEXP (a, 1));
     }
 
   if (a_op1 == NULL_RTX)
     goto fail;
 
   /* Ensure the cond is of form: x = (y op z) : y */
-  a_op0 = get_base_reg (XEXP (a, 0));
+  a_op0 = get_base_reg_or_const_int (XEXP (a, 0));
   if (!(a_op0 && rtx_equal_p (a_op0, b)))
     goto fail;
 
   start_sequence ();
 
-  target = gen_reg_rtx (GET_MODE (XEXP (a, op != AND)));
+  /* XEXP (a, op != AND) might be a CONST_INT and thus VOIDmode.  In
+     that case fall back to WORD_MODE and hope for the best.  The worry
+     is asserts in the various move paths tripping on a mode mismatch.  */
+  tmp_mode = GET_MODE (XEXP (a, op != AND));
+  tmp_mode = tmp_mode == VOIDmode ? word_mode : tmp_mode;
+  target = gen_reg_rtx (tmp_mode);
 
   /* AND requires !cond, instead we swap ops around.  */
   target = noce_emit_cmove (if_info, target, code,
diff --git a/gcc/testsuite/gcc.target/riscv/pr125731.c b/gcc/testsuite/gcc.target/riscv/pr125731.c
new file mode 100644
index 00000000000..d041f900d50
--- /dev/null
+++ b/gcc/testsuite/gcc.target/riscv/pr125731.c
@@ -0,0 +1,18 @@
+/* { dg-do compile } */
+/* { dg-options "-march=rv64gcb_zicond -mabi=lp64d" { target rv64 } } */
+/* { dg-options "-march=rv32gcb_zicond -mabi=ilp32" { target rv32 } } */
+/* { dg-skip-if "" { *-*-* } { "-O0" "-Og"} } */
+
+long fun_not1 (int a, long b)
+{
+    if (!(a & 1))
+        b ^= 8;
+    return b;
+}
+
+
+/* { dg-final { scan-assembler-times "\\tczero" 1 } } */
+/* { dg-final { scan-assembler-not "\\tbeq\\t" } } */
+/* { dg-final { scan-assembler-not "\\tbne\\t" } } */
+
+
