diff --git a/gcc/ifcvt.cc b/gcc/ifcvt.cc
index 1df5d54efa36..d58d4dfba98a 100644
--- a/gcc/ifcvt.cc
+++ b/gcc/ifcvt.cc
@@ -781,6 +781,7 @@ static bool noce_try_ifelse_collapse (struct noce_if_info *);
 static bool noce_try_store_flag (struct noce_if_info *);
 static bool noce_try_addcc (struct noce_if_info *);
 static bool noce_try_store_flag_constants (struct noce_if_info *);
+static bool noce_try_shifted_store_flag (struct noce_if_info *);
 static bool noce_try_store_flag_mask (struct noce_if_info *);
 static rtx noce_emit_cmove (struct noce_if_info *, rtx, enum rtx_code, rtx,
 			    rtx, rtx, rtx, rtx = NULL, rtx = NULL);
@@ -1558,6 +1559,155 @@ noce_try_inverse_constants (struct noce_if_info *if_info)
   return false;
 }
 
+/*  Check if OP is supported by conditional zero based if conversion,
+    returning TRUE if satisfied otherwise FALSE.
+
+    OP is the operation to check.  */
+
+static bool
+noce_cond_zero_binary_op_supported (rtx op)
+{
+  enum rtx_code opcode = GET_CODE (op);
+
+  if (opcode == PLUS || opcode == MINUS || opcode == IOR || opcode == XOR
+      || opcode == ASHIFT || opcode == ASHIFTRT || opcode == LSHIFTRT
+      || opcode == ROTATE || opcode == ROTATERT || opcode == AND)
+    return true;
+
+  return false;
+}
+
+/* Convert "if (test) x = a; else x = a OP c  " where c is 2^n.
+
+   We can shift the output of a set flag insn left to generate 2^n
+   cheaply, then apply OP unconditionally.  This works even if the
+   target does not have a conditional zero idiom.  On targets such
+   as RISC-V this sequence may also encode better.
+
+   This is based on noce_try_store_flag_constants.  */
+
+static bool
+noce_try_shifted_store_flag (struct noce_if_info *if_info)
+{
+  rtx target;
+  rtx_insn *seq;
+  bool reversep;
+  HOST_WIDE_INT itrue, ifalse;
+  int normalize;
+  bool can_reverse;
+  machine_mode mode = GET_MODE (if_info->x);
+  rtx common = NULL_RTX;
+  rtx_code code = UNKNOWN;
+
+  if (STORE_FLAG_VALUE != 1)
+    return false;
+
+  rtx a = if_info->a;
+  rtx b = if_info->b;
+
+  /* Basically any binary operator is allowed and we want
+     A in one arm and A OP C in the other where C is 2^n.  */
+  if (noce_cond_zero_binary_op_supported (a)
+      && GET_CODE (b) == REG
+      && rtx_equal_p (XEXP (a, 0), b)
+      && CONST_INT_P (XEXP (a, 1))
+      && pow2p_hwi (INTVAL (XEXP (a, 1))))
+    {
+      code = GET_CODE (a);
+      common = XEXP (a, 0);
+      a = XEXP (a, 1);
+      b = CONST0_RTX (GET_MODE (common));
+    }
+
+  if (noce_cond_zero_binary_op_supported (b)
+      && GET_CODE (a) == REG
+      && rtx_equal_p (XEXP (b, 0), a)
+      && CONST_INT_P (XEXP (b, 1))
+      && pow2p_hwi (INTVAL (XEXP (b, 1))))
+    {
+      code = GET_CODE (b);
+      common = XEXP (b, 0);
+      b = XEXP (b, 1);
+      a = CONST0_RTX (GET_MODE (common));
+    }
+
+  /* If CODE hasn't been reset, then the basic expression form was wrong and
+     we can not optimize.  We also can not optimize with nontrivial blocks.  */
+  if (code == UNKNOWN || !noce_simple_bbs (if_info))
+    return false;
+
+  ifalse = INTVAL (a);
+  itrue = INTVAL (b);
+
+  can_reverse = noce_reversed_cond_code (if_info) != UNKNOWN;
+  reversep = false;
+  /* Is this (cond) ? 2^n : 0?  */
+  if (ifalse == 0 && pow2p_hwi (itrue))
+    ;
+  /* Is this (cond) ? 0 : 2^n?  */
+  else if (itrue == 0 && pow2p_hwi (ifalse) && can_reverse)
+    reversep = true;
+  else
+    return false;
+
+  if (reversep)
+    std::swap (itrue, ifalse);
+
+  start_sequence ();
+
+  /* If X and COMMON are the same, then we're going to need a temporary.  */
+  if (common && rtx_equal_p (common, if_info->x))
+    {
+      common = gen_reg_rtx (mode);
+      noce_emit_move_insn (common, if_info->x);
+    }
+
+  target = noce_emit_store_flag (if_info, if_info->x, reversep, 1);
+  if (!target)
+    {
+      end_sequence ();
+      return false;
+    }
+
+  /* Right now TARGET has the value 1/0.  A left shift will produce the
+     target value.  Exception if the target is 1, we're already there.  */
+  if (itrue != 1)
+    target = expand_simple_binop (mode, ASHIFT, target,
+				  GEN_INT (exact_log2 (itrue)),
+				  NULL_RTX, 0, OPTAB_WIDEN);
+
+  if (!target)
+    {
+      end_sequence ();
+      return false;
+    }
+
+  /* Now we just need to apply the code to COMMON and TARGET.  */
+  target = expand_simple_binop (mode, code,
+				common, target,
+				NULL_RTX, 0, OPTAB_WIDEN);
+
+  if (!target)
+    {
+      end_sequence ();
+      return false;
+    }
+
+  /* If necessary, store the result into the proper destination.  */
+  if (target != if_info->x)
+    noce_emit_move_insn (if_info->x, target);
+
+  seq = end_ifcvt_sequence (if_info);
+  if (!seq || !targetm.noce_conversion_profitable_p (seq, if_info))
+    return false;
+
+  emit_insn_before_setloc (seq, if_info->jump,
+			   INSN_LOCATION (if_info->insn_a));
+  if_info->transform_name = "noce_try_shifted_store_flag";
+
+  return true;
+}
+
 
 /* Convert "if (test) x = a; else x = b", for A and B constant.
    Also allow A = y + c1, B = y + c2, with a common y between A
@@ -3167,34 +3317,18 @@ noce_try_sign_mask (struct noce_if_info *if_info)
   return true;
 }
 
-/*  Check if OP is supported by conditional zero based if conversion,
-    returning TRUE if satisfied otherwise FALSE.
-
-    OP is the operation to check.  */
-
-static bool
-noce_cond_zero_binary_op_supported (rtx op)
-{
-  enum rtx_code opcode = GET_CODE (op);
-
-  if (opcode == PLUS || opcode == MINUS || opcode == IOR || opcode == XOR
-      || opcode == ASHIFT || opcode == ASHIFTRT || opcode == LSHIFTRT
-      || opcode == ROTATE || opcode == ROTATERT || opcode == AND)
-    return true;
-
-  return false;
-}
-
 /*  Helper function to return REG itself,
     otherwise NULL_RTX for other RTX_CODE.  */
 
 static rtx
-get_base_reg (rtx exp)
+get_base_reg_or_const_int (rtx exp)
 {
   if (REG_P (exp))
     return exp;
   else if (SUBREG_P (exp))
     return SUBREG_REG (exp);
+  else if (CONST_INT_P (exp))
+    return exp;
 
   return NULL_RTX;
 }
@@ -3232,6 +3366,7 @@ noce_try_cond_arith (struct noce_if_info *if_info)
   rtx_insn *seq;
   rtx_code op;
   machine_mode mode = GET_MODE (if_info->x);
+  machine_mode tmp_mode;
 
   /* Scalar integral modes are only supported here.
      Could support scalar floating point but that
@@ -3269,24 +3404,29 @@ noce_try_cond_arith (struct noce_if_info *if_info)
   op = GET_CODE (a);
 
   /* Canonicalize x = (z op y) : y to x = (y op z) : y */
-  a_op1 = get_base_reg (XEXP (a, 1));
+  a_op1 = get_base_reg_or_const_int (XEXP (a, 1));
   if (a_op1 && rtx_equal_p (a_op1, b) && COMMUTATIVE_ARITH_P (a))
     {
       std::swap (XEXP (a, 0), XEXP (a, 1));
-      a_op1 = get_base_reg (XEXP (a, 1));
+      a_op1 = get_base_reg_or_const_int (XEXP (a, 1));
     }
 
   if (a_op1 == NULL_RTX)
     goto fail;
 
   /* Ensure the cond is of form: x = (y op z) : y */
-  a_op0 = get_base_reg (XEXP (a, 0));
+  a_op0 = get_base_reg_or_const_int (XEXP (a, 0));
   if (!(a_op0 && rtx_equal_p (a_op0, b)))
     goto fail;
 
   start_sequence ();
 
-  target = gen_reg_rtx (GET_MODE (XEXP (a, op != AND)));
+  /* XEXP (a, op != AND) might be a CONST_INT and thus VOIDmode.  In
+     that case fall back to WORD_MODE and hope for the best.  The worry
+     is asserts in the various move paths tripping on a mode mismatch.  */
+  tmp_mode = GET_MODE (XEXP (a, op != AND));
+  tmp_mode = tmp_mode == VOIDmode ? word_mode : tmp_mode;
+  target = gen_reg_rtx (tmp_mode);
 
   /* AND requires !cond, instead we swap ops around.  */
   target = noce_emit_cmove (if_info, target, code,
@@ -4550,6 +4690,9 @@ noce_process_if_block (struct noce_if_info *if_info)
   if (!targetm.have_conditional_execution ()
       && noce_try_store_flag_constants (if_info))
     goto success;
+  if (!targetm.have_conditional_execution ()
+      && noce_try_shifted_store_flag (if_info))
+    goto success;
   if (noce_try_sign_bit_splat (if_info))
     goto success;
   if (!targetm.have_conditional_execution ()
