
diff --git a/gcc/simplify-rtx.cc b/gcc/simplify-rtx.cc
index 8f0f16c865d1..8688d5361ff2 100644
--- a/gcc/simplify-rtx.cc
+++ b/gcc/simplify-rtx.cc
@@ -3696,6 +3696,17 @@ simplify_context::simplify_binary_operation_1 (rtx_code code,
 	      && !side_effects_p (XEXP (op0, 0)))
 	    return trueop1;
 
+	  /* Generic code elsewhere turns irrelevant bits in C1.  But
+	     if we turn them back on and the result is a constant with
+	     a single bit off, that's likely better as many targets
+	     have single bit manipulation instructions.  */
+	  if (exact_log2 (~(c1 | c2)) >= 0)
+	    {
+	      rtx temp = gen_rtx_AND (mode, XEXP (op0, 0), GEN_INT (c1 | c2));
+	      temp = gen_rtx_IOR (mode, temp, trueop1);
+	      return temp;
+	    }
+
 	  /* If (C1|C2) == ~0 then (X&C1)|C2 becomes X|C2.  */
 	  if (((c1|c2) & mask) == mask)
 	    return simplify_gen_binary (IOR, mode, XEXP (op0, 0), op1);
