diff --git a/gcc/testsuite/gcc.target/riscv/zbb-sext.c b/gcc/testsuite/gcc.target/riscv/zbb-sext.c
index 1ad3e714371..bcab8140468 100644
--- a/gcc/testsuite/gcc.target/riscv/zbb-sext.c
+++ b/gcc/testsuite/gcc.target/riscv/zbb-sext.c
@@ -1,7 +1,7 @@
 /* { dg-do compile } */
 /* { dg-options "-O2 -march=rv64gc_zbb -mabi=lp64" } */
 
-#include<stdint.h>
+#include <stdint-gcc.h>
 
 int8_t foo1(uint8_t a) {
     return a;
