diff --git a/gcc/match.pd b/gcc/match.pd
index 492d88514fce..ea8889644670 100644
--- a/gcc/match.pd
+++ b/gcc/match.pd
@@ -12180,7 +12180,10 @@ and,
 (simplify
   (bit_ior:c (lshift @0 INTEGER_CST@1)
 	     (bit_xor (rshift @2 INTEGER_CST@3) INTEGER_CST@4))
-   (if (((~((HOST_WIDE_INT_1U << tree_to_uhwi (@1)) - 1)) & tree_to_uhwi (@4)) == 0
+   (if (tree_fits_uhwi_p (@1)
+	&& tree_fits_uhwi_p (@3)
+	&& tree_fits_uhwi_p (@4)
+	&& ((~((HOST_WIDE_INT_1U << tree_to_uhwi (@1)) - 1)) & tree_to_uhwi (@4)) == 0
         && (tree_to_uhwi (@1) + tree_to_uhwi (@3)) == TYPE_PRECISION (type)
         && TYPE_UNSIGNED (type)
         && @0 == @2)
@@ -12191,7 +12194,10 @@ and,
 (simplify
   (bit_ior:c (bit_xor (lshift @0 INTEGER_CST@1) INTEGER_CST@2)
 	     (rshift @3 INTEGER_CST@4))
-   (if ((((((HOST_WIDE_INT_1U << tree_to_uhwi (@1)) - 1)) & tree_to_uhwi (@2)) == 0)
+   (if (tree_fits_uhwi_p (@1)
+	&& tree_fits_uhwi_p (@2)
+	&& tree_fits_uhwi_p (@4)
+	&& (((((HOST_WIDE_INT_1U << tree_to_uhwi (@1)) - 1)) & tree_to_uhwi (@2)) == 0)
         && (tree_to_uhwi (@1) + tree_to_uhwi (@4)) == TYPE_PRECISION (type)
         && TYPE_UNSIGNED (type)
         && @0 == @3)
diff --git a/gcc/testsuite/gcc.dg/torture/pr123530.c b/gcc/testsuite/gcc.dg/torture/pr123530.c
new file mode 100644
index 000000000000..1012270098bf
--- /dev/null
+++ b/gcc/testsuite/gcc.dg/torture/pr123530.c
@@ -0,0 +1,6 @@
+/* { dg-do compile } */
+int a, b, c;
+void d() {
+  int e = -1;
+  (c >> e | ~(b << 1) ^ 1) & a;
+}
