diff --git a/gcc/tree-vect-stmts.cc b/gcc/tree-vect-stmts.cc
index d642d3c257f..3dd8a284577 100644
--- a/gcc/tree-vect-stmts.cc
+++ b/gcc/tree-vect-stmts.cc
@@ -6481,6 +6481,10 @@ vectorizable_operation (vec_info *vinfo,
   scalar_dest = gimple_assign_lhs (stmt);
   vectype_out = STMT_VINFO_VECTYPE (stmt_info);
 
+  /* STMT may have already been vectorized.  */
+  if (VECTOR_TYPE_P (TREE_TYPE (scalar_dest)))
+    return false;
+
   /* Most operations cannot handle bit-precision types without extra
      truncations.  */
   bool mask_op_p = VECTOR_BOOLEAN_TYPE_P (vectype_out);
