diff --git a/gcc/match.pd b/gcc/match.pd
index 6aaf80eee7d5..90a4442fea80 100644
--- a/gcc/match.pd
+++ b/gcc/match.pd
@@ -12028,3 +12028,17 @@ and,
 (simplify
  (IFN_VEC_SHL_INSERT (vec_duplicate@1 @0) @0)
   @1)
+
+/* Be extra careful with the shift count.  Don't transform a negative count
+   or a count larger than the object's size.  */
+(simplify
+ (bit_and (rshift SSA_NAME@0 INTEGER_CST@1) integer_onep)
+ (if (tree_int_cst_sgn (@1) >= 0
+      && wi::ltu_p (wi::to_wide (@1), TYPE_PRECISION (TREE_TYPE (@0)))
+      && (fold_before_rtl_expansion_p ()
+          || !flag_tree_vrp
+          || optimize_debug)
+      && !BYTES_BIG_ENDIAN)
+  (convert (BIT_FIELD_REF:boolean_type_node @0
+	    { build_one_cst (bitsizetype); }
+	    (convert:bitsizetype @1)))))
diff --git a/gcc/testsuite/gcc.target/riscv/pr32648.c b/gcc/testsuite/gcc.target/riscv/pr32648.c
new file mode 100644
index 000000000000..f4f5975bde5a
--- /dev/null
+++ b/gcc/testsuite/gcc.target/riscv/pr32648.c
@@ -0,0 +1,78 @@
+/* { dg-do compile } */
+/* { dg-options "-O2 -fdump-tree-optimized" } */
+
+#define F1(OP,NAME,TYPE,RETTYPE,NT,NRT) RETTYPE f##NAME##NT##NRT (TYPE a) { TYPE b5 = (a & 0x20)>>5; TYPE b3 = (a & 0x08)>>3; return b5 OP b3; }
+#define F2(OP,NAME,TYPE,RETTYPE,NT,NRT) RETTYPE f2##NAME##NT##NRT (TYPE a) { TYPE b5 = (a & 0x20)>>5; TYPE b3 = (a & 0x08)>>3; return (b5 OP ~b3) & 0x1; }
+
+#define F(OP,NAME) \
+F1 (OP, NAME, unsigned char, unsigned char, uc, uc)  \
+F2 (OP, NAME, signed char, unsigned char, sc, uc) \
+F1 (OP, NAME, unsigned short, unsigned char, us, uc) \
+F2 (OP, NAME, signed short, unsigned char, ss, uc) \
+F1 (OP, NAME, unsigned int, unsigned char, ui, uc) \
+F2 (OP, NAME, signed int, unsigned char, si, uc) \
+F1 (OP, NAME, unsigned long, unsigned char, ul, uc) \
+F2 (OP, NAME, signed long, unsigned char, sl, uc) \
+F1 (OP, NAME, unsigned char, signed char, uc, sc) \
+F2 (OP, NAME, signed char, signed char, sc, sc) \
+F1 (OP, NAME, unsigned short, signed char, us, sc)\
+F2 (OP, NAME, signed short, signed char, ss, sc)\
+F1 (OP, NAME, unsigned int, signed char, ui, sc)\
+F2 (OP, NAME, signed int, signed char, si, sc)\
+F1 (OP, NAME, unsigned long, signed char, ul, sc)\
+F2 (OP, NAME, signed long, signed char, sl, sc)\
+F1 (OP, NAME, unsigned char, unsigned short, uc, us)\
+F2 (OP, NAME, signed char, unsigned short, sc, us)\
+F1 (OP, NAME, unsigned short, unsigned short, us, us)\
+F2 (OP, NAME, signed short, unsigned short, ss, us)\
+F1 (OP, NAME, unsigned int, unsigned short, ui, us)\
+F2 (OP, NAME, signed int, unsigned short, si, us)\
+F1 (OP, NAME, unsigned long, unsigned short, ul, us)\
+F2 (OP, NAME, signed long, unsigned short, sl, us)\
+F1 (OP, NAME, unsigned char, signed short, uc, ss)\
+F2 (OP, NAME, signed char, signed short, sc, ss)\
+F1 (OP, NAME, unsigned short, signed short, us, ss)\
+F2 (OP, NAME, signed short, signed short, ss, ss)\
+F1 (OP, NAME, unsigned int, signed short, ui, ss)\
+F2 (OP, NAME, signed int, signed short, si, ss)\
+F1 (OP, NAME, unsigned long, signed short, ul, ss)\
+F2 (OP, NAME, signed long, signed short, sl, ss)\
+F1 (OP, NAME, unsigned char, unsigned int, uc, ui)\
+F2 (OP, NAME, signed char, unsigned int, sc, ui)\
+F1 (OP, NAME, unsigned short, unsigned int, us, ui)\
+F2 (OP, NAME, signed short, unsigned int, ss, ui)\
+F1 (OP, NAME, unsigned int, unsigned int, ui, ui)\
+F2 (OP, NAME, signed int, unsigned int, si, ui)\
+F1 (OP, NAME, unsigned long, unsigned int, ul, ui)\
+F2 (OP, NAME, signed long, unsigned int, sl, ui)\
+F1 (OP, NAME, unsigned char, signed int, uc, si)\
+F2 (OP, NAME, signed char, signed int, sc, si)\
+F1 (OP, NAME, unsigned short, signed int, us, si)\
+F2 (OP, NAME, signed short, signed int, ss, si)\
+F1 (OP, NAME, unsigned int, signed int, ui, si)\
+F2 (OP, NAME, signed int, signed int, si, si)\
+F1 (OP, NAME, unsigned long, signed int, ul, si)\
+F2 (OP, NAME, signed long, signed int, sl, si)\
+F1 (OP, NAME, unsigned char, unsigned long, uc, ul)\
+F2 (OP, NAME, signed char, unsigned long, sc, ul)\
+F1 (OP, NAME, unsigned short, unsigned long, us, ul)\
+F2 (OP, NAME, signed short, unsigned long, ss, ul)\
+F1 (OP, NAME, unsigned int, unsigned long, ii, ul)\
+F2 (OP, NAME, signed int, unsigned long, si, ul)\
+F1 (OP, NAME, unsigned long, unsigned long, ul, ul)\
+F2 (OP, NAME, signed long, unsigned long, sl, ul)\
+F1 (OP, NAME, unsigned char, signed long, uc, sl)\
+F2 (OP, NAME, signed char, signed long, sc, sl)\
+F1 (OP, NAME, unsigned short, signed long, us, sl)\
+F2 (OP, NAME, signed short, signed long, ss, sl)\
+F1 (OP, NAME, unsigned int, signed long, ui, sl)\
+F2 (OP, NAME, signed int, signed long, si, sl)\
+F1 (OP, NAME, unsigned long, signed long, ul, sl)\
+F2 (OP, NAME, signed long, signed long, sl, sl)
+
+
+F(&,_and)
+F(|,_or)
+F(^,_xor)
+
+/* { dg-final { scan-tree-dump-times "BIT_FIELD_REF" 384 "optimized"} } */
diff --git a/gcc/testsuite/gcc.target/riscv/pr64345.c b/gcc/testsuite/gcc.target/riscv/pr64345.c
index 8ca4e2411ee9..21ddea9b1d36 100644
--- a/gcc/testsuite/gcc.target/riscv/pr64345.c
+++ b/gcc/testsuite/gcc.target/riscv/pr64345.c
@@ -20,8 +20,7 @@ int test6 (int vi) { return vi - (((vi >> 6) & 0x01) << 1) - 1; }
 
 
 /* { dg-final { scan-assembler-times "\\tbexti" 5 } } */
-/* { dg-final { scan-assembler-times "\\txori" 3 } } */
-/* { dg-final { scan-assembler-times "\\tnot" 1 } } */
+/* { dg-final { scan-assembler-times "\\txori" 4 } } */
 /* { dg-final { scan-assembler-times "\\tsrli" 2 } } */
 /* { dg-final { scan-assembler-times "\\tandi" 2 } } */
 /* { dg-final { scan-assembler-times "\\tsub" 2 } } */
