diff --git a/gcc/config/rx/rx.cc b/gcc/config/rx/rx.cc
index 0754e286552..a2cffa0e0af 100644
--- a/gcc/config/rx/rx.cc
+++ b/gcc/config/rx/rx.cc
@@ -2791,6 +2791,7 @@ rx_override_options_after_change (void)
 
       first_time = FALSE;
     }
+#if 0
   else
     {
       /* Alert the user if they are changing the optimization options
@@ -2799,6 +2800,7 @@ rx_override_options_after_change (void)
 	  && !flag_finite_math_only)
 	warning (0, "RX FPU instructions do not support NaNs and infinities");
     }
+#endif
 }
 
 static void
