diff --git a/gcc/testsuite/gcc.c-torture/execute/20001229-1.c b/gcc/testsuite/gcc.c-torture/execute/20001229-1.c
index 3bf333dd5a8..a235fdad138 100644
--- a/gcc/testsuite/gcc.c-torture/execute/20001229-1.c
+++ b/gcc/testsuite/gcc.c-torture/execute/20001229-1.c
@@ -14,6 +14,8 @@ void exit (int);
 #include <asm/sysinfo.h>
 #include <asm/unistd.h>
 
+int syscall (int, ...);
+
 static inline int
 setsysinfo(unsigned long op, void *buffer, unsigned long size,
            int *start, void *arg, unsigned long flag)
