diff --git a/gcc/lra-constraints.cc b/gcc/lra-constraints.cc
index b15299791745..6990e4d1e6b4 100644
--- a/gcc/lra-constraints.cc
+++ b/gcc/lra-constraints.cc
@@ -523,6 +523,11 @@ update_equiv (int regno)
 {
   rtx x;
 
+  /* If REGNO is beyond the length of the equivalence array structure,
+     then there's nothing to update.  */
+  if (regno >= ira_reg_equiv_len)
+    return;
+
   if ((x = ira_reg_equiv[regno].memory) != NULL_RTX)
     ira_reg_equiv[regno].memory
       = simplify_replace_fn_rtx (x, NULL_RTX, loc_equivalence_callback,
