diff --git a/gcc/match.pd b/gcc/match.pd
index 3cd9ab1e9b04..6f0d3acd2bc9 100644
--- a/gcc/match.pd
+++ b/gcc/match.pd
@@ -11994,3 +11994,12 @@ and,
 (simplify
  (IFN_VEC_SHL_INSERT (vec_duplicate@1 @0) @0)
   @1)
+
+(simplify
+  (bit_ior:c (lshift @0 INTEGER_CST@1)
+             (bit_xor (rshift @2 INTEGER_CST@3) INTEGER_CST@4))
+  (if (((~((HOST_WIDE_INT_1U << tree_to_uhwi (@1)) - 1)) & tree_to_uhwi (@4)) == 0
+      && (tree_to_uhwi (@1) + tree_to_uhwi (@3)) == TYPE_PRECISION (type)
+      && TYPE_UNSIGNED (type)
+      && @0 == @2)
+    (bit_xor (bit_ior (lshift @0 @1) (rshift @2 @3)) @4)))
