diff --git a/gcc/config/elfos.h b/gcc/config/elfos.h
index acb376283cf..09dd4c89985 100644
--- a/gcc/config/elfos.h
+++ b/gcc/config/elfos.h
@@ -214,8 +214,8 @@ see the files COPYING3 and COPYING.RUNTIME respectively.  If not, see
    The definitions say how to change sections to the .init and .fini
    sections.  This is the same for all known svr4 assemblers.  */
 
-#define INIT_SECTION_ASM_OP	"\t.section\t.init"
-#define FINI_SECTION_ASM_OP	"\t.section\t.fini"
+#define INIT_SECTION_ASM_OP	"\t.section\t.init,\"ax\""
+#define FINI_SECTION_ASM_OP	"\t.section\t.fini,\"ax\""
 
 /* Output assembly directive to move to the beginning of current section.  */
 #ifdef HAVE_GAS_SUBSECTION_ORDERING
