diff --git a/gcc/tree-ssa-threadupdate.c b/gcc/tree-ssa-threadupdate.c
index 2b9b8f81274..56106e4ff6a 100644
--- a/gcc/tree-ssa-threadupdate.c
+++ b/gcc/tree-ssa-threadupdate.c
@@ -2784,8 +2784,7 @@ jt_path_registry::cancel_invalid_paths (vec<jump_thread_edge *> &path)
       // The first entry represents the block with an outgoing edge
       // that we will redirect to the jump threading path.  Thus we
       // don't care about that block's loop father.
-      if ((i > 0 && e->src->loop_father != loop)
-	  || e->dest->loop_father != loop)
+      if ((i > 0 && e->src->loop_father != loop))
 	path_crosses_loops = true;
 
       if (flag_checking && !m_backedge_threads)

