diff --git a/gcc/ifcvt.cc b/gcc/ifcvt.cc
index 0858b3d2f6d6..a8726e796839 100644
--- a/gcc/ifcvt.cc
+++ b/gcc/ifcvt.cc
@@ -4264,6 +4264,7 @@ noce_process_if_block (struct noce_if_info *if_info)
   unsigned old_cost = if_info->original_cost;
   if (!else_bb
       && HAVE_conditional_move
+      && ifcvt_after_combine
       && bb_ok_for_noce_convert_multiple_sets (then_bb, &potential_cost))
     {
       /* Temporarily set the original costs to what we estimated so
