diff --git a/gcc/config/riscv/riscv.cc b/gcc/config/riscv/riscv.cc
index 949616a63d3c..a85a2035a8b7 100644
--- a/gcc/config/riscv/riscv.cc
+++ b/gcc/config/riscv/riscv.cc
@@ -13793,7 +13793,9 @@ riscv_emit_mode_set (int entity, int mode, int prev_mode,
   switch (entity)
     {
     case RISCV_VXRM:
-      if (mode != VXRM_MODE_NONE && mode != prev_mode)
+      if (mode != VXRM_MODE_NONE
+	  && mode != VXRM_MODE_CLOBBER
+	  && mode != prev_mode)
 	emit_insn (gen_vxrmsi (gen_int_mode (mode, SImode)));
       break;
     case RISCV_FRM:
diff --git a/gcc/config/riscv/vector.md b/gcc/config/riscv/vector.md
index 8e20aac9bb8e..6bf01338117c 100644
--- a/gcc/config/riscv/vector.md
+++ b/gcc/config/riscv/vector.md
@@ -1003,7 +1003,7 @@ (define_attr "avl_type_idx" ""
 
 ;; Defines rounding mode of an fixed-point operation.
 
-(define_attr "vxrm_mode" "rnu,rne,rdn,rod,none"
+(define_attr "vxrm_mode" "rnu,rne,rdn,rod,clobber,none"
   (cond [(eq_attr "type" "vaalu,vsmul,vsshift,vnclip")
 	 (cond
 	   [(match_test "INTVAL (operands[9]) == riscv_vector::VXRM_RNU")
@@ -1017,7 +1017,9 @@ (define_attr "vxrm_mode" "rnu,rne,rdn,rod,none"
 
 	    (match_test "INTVAL (operands[9]) == riscv_vector::VXRM_ROD")
 	    (const_string "rod")]
-	   (const_string "none"))]
+	   (const_string "none"))
+	 (match_test "CALL_P (insn)")
+	 (const_string "clobber")]
         (const_string "none")))
 
 ;; Defines rounding mode of an floating-point operation.
diff --git a/gcc/testsuite/gcc.target/riscv/rvv/base/pr123626.c b/gcc/testsuite/gcc.target/riscv/rvv/base/pr123626.c
new file mode 100644
index 000000000000..4816f03c109a
--- /dev/null
+++ b/gcc/testsuite/gcc.target/riscv/rvv/base/pr123626.c
@@ -0,0 +1,20 @@
+short a;
+long long b;
+char c[3][3][17];
+_Bool d;
+
+int main() {
+  for (long g=0; g<3; ++g)
+    for (long h=0; h<3; ++h)
+      for (long i=0; i<17; ++i)
+        c[g][h][i] = 2;
+
+  for (char g=0; g<3; g-=13)
+    for (short j=3; j<d+21; j+=2)
+      a += 2;
+
+  b = (int)a;
+  if (b != 180)
+    __builtin_abort ();
+  __builtin_exit (0);
+}
