diff --git a/gcc/cselib.cc b/gcc/cselib.cc
index 930357409bc5..07d87980b167 100644
--- a/gcc/cselib.cc
+++ b/gcc/cselib.cc
@@ -1247,7 +1247,8 @@ cselib_redundant_set_p (rtx set)
     src = XEXP (src, 0);
 
   if (MEM_P (src)
-      && rtx_equal_for_cselib_1 (dest_addr, XEXP (src, 0), GET_MODE (dest), 0))
+      && rtx_equal_for_cselib_1 (dest_addr, XEXP (src, 0), GET_MODE (dest), 0)
+      && !(MEM_VOLATILE_P (src) || MEM_VOLATILE_P (dest)))
     return mems_same_for_tbaa_p (src, dest);
 
   return false;
