diff --git a/gcc/ext-dce.cc b/gcc/ext-dce.cc
index 67ec92a42878..836d4027a4d1 100644
--- a/gcc/ext-dce.cc
+++ b/gcc/ext-dce.cc
@@ -523,7 +523,7 @@ carry_backpropagate (unsigned HOST_WIDE_INT mask, enum rtx_code code, rtx x)
     case ASHIFT:
       if (CONST_INT_P (XEXP (x, 1))
 	  && UINTVAL (XEXP (x, 1)) < GET_MODE_BITSIZE (smode))
-	return (HOST_WIDE_INT) mask >> INTVAL (XEXP (x, 1));
+	return mask >> INTVAL (XEXP (x, 1));
       return (HOST_WIDE_INT_UC (2) << floor_log2 (mask)) - 1;
 
     /* We propagate for the shifted operand, but not the shift
