diff --git a/gcc/ifcvt.cc b/gcc/ifcvt.cc
index 37e757461934..1df5d54efa36 100644
--- a/gcc/ifcvt.cc
+++ b/gcc/ifcvt.cc
@@ -1797,12 +1797,12 @@ noce_try_store_flag_logical (struct noce_if_info *if_info)
   if (!noce_simple_bbs (if_info))
     return false;
 
-  bool swapped = false;
+  bool swapped = STORE_FLAG_VALUE == -1;
   if (a == CONSTM1_RTX (GET_MODE (dest))
       && if_info->rev_cond)
     {
       std::swap (a, b);
-      swapped = true;
+      swapped = !swapped;
     }
 
   if (b != CONSTM1_RTX (GET_MODE (dest)))
diff --git a/gcc/testsuite/gcc.dg/torture/pr126136.c b/gcc/testsuite/gcc.dg/torture/pr126136.c
new file mode 100644
index 000000000000..cc270f51ca2d
--- /dev/null
+++ b/gcc/testsuite/gcc.dg/torture/pr126136.c
@@ -0,0 +1,24 @@
+/* { dg-do run } */
+
+int i;
+
+__attribute__ ((noipa,noclone,noinline))
+char foo(char val)
+{
+  i++;
+  if (i > 1)
+    return -1;
+  else
+    return val;
+}
+
+int
+main(void)
+{
+  i = 0;
+  if (foo (10) != 10)
+    __builtin_abort ();
+  i += 2;
+  if (foo (10) != -1)
+    __builtin_abort ();
+}
