diff --git a/gcc/config/riscv/riscv.cc b/gcc/config/riscv/riscv.cc
index 63404d3d514..8880b199d41 100644
--- a/gcc/config/riscv/riscv.cc
+++ b/gcc/config/riscv/riscv.cc
@@ -12040,6 +12040,14 @@ riscv_option_override (void)
 		       param_cycle_accurate_model,
 		       0);
 
+  /* Cache lines of 64 bytes are common these days, so let's get a sensible
+     value for the interference size.  Technically this can leak and cause
+     sizes of structures to change, but consensus is anything using the
+     value to size fields within a structure is broken.  Any 2^n byte value
+     is functionally correct, but may not be performant.  */
+  SET_OPTION_IF_UNSET (&global_options, &global_options_set,
+		       param_destruct_interfere_size, 0);
+
   /* Function to allocate machine-dependent function status.  */
   init_machine_status = &riscv_init_machine_status;
 
