diff --git a/gcc/config/riscv/riscv-vsetvl.cc b/gcc/config/riscv/riscv-vsetvl.cc
index 3586d0cdcc24..97662153e8e5 100644
--- a/gcc/config/riscv/riscv-vsetvl.cc
+++ b/gcc/config/riscv/riscv-vsetvl.cc
@@ -2998,6 +2998,11 @@ pre_vsetvl::earliest_fuse_vsetvl_info (int iter)
 	      || eg->probability == profile_probability::never ()
 	      || src_block_info.probability
 		   == profile_probability::uninitialized ()
+	      /* Hoisting across an edge that is unlikely to be executed can be
+		 actively harmful.  A better option would be to sum the probabilities
+		 of all the edges we could hoist through.  */
+	      || eg->probability <= profile_probability::guessed_always () / 10
+	      || eg->probability == profile_probability::never ()
 	      /* When multiple set bits in earliest edge, such edge may
 		 have infinite loop in preds or succs or multiple conflict
 		 vsetvl expression which make such edge is unrelated.  We
