diff --git a/gcc/testsuite/gcc.target/riscv/rvv/autovec/pr122445.c b/gcc/testsuite/gcc.target/riscv/rvv/autovec/pr122445.c
index 47368684faa1..cd046647656a 100644
--- a/gcc/testsuite/gcc.target/riscv/rvv/autovec/pr122445.c
+++ b/gcc/testsuite/gcc.target/riscv/rvv/autovec/pr122445.c
@@ -22,5 +22,5 @@ x264_add4x4_idct (uint8_t *p_dst, int16_t d[16])
     }
 }
 
-/* { dg-final { scan-rtl-dump "Propagating AVL: \\(const_int 4" "avlprop" } } */
+/* { dg-final { scan-rtl-dump "Propagating AVL: \\(const_int 8" "avlprop" } } */
 /* { dg-final { scan-rtl-dump-not "Propagating AVL: \\(const_int 1" "avlprop" } } */
