diff --git a/gcc/config/riscv/riscv.cc b/gcc/config/riscv/riscv.cc
index f82c354101d8..b3a1f4247733 100644
--- a/gcc/config/riscv/riscv.cc
+++ b/gcc/config/riscv/riscv.cc
@@ -15984,6 +15984,55 @@ synthesize_and (rtx operands[3])
       return true;
     }
 
+  /* If it's a run of 1s in the high and low bits and the popcount of the
+     run of 1s is <= 11, 16 or 32, then we can rotate all the bits we
+     want to preserve into the low bits, andi or sext.[hw], then
+     rotate the bits back into position. 
+
+     Zbb is needed for all these sequences.  Some may need Zba as well.  */
+  if (TARGET_64BIT
+      && TARGET_ZBB
+      && budget >= 3
+      && consecutive_bits_operand (GEN_INT (~t), word_mode)
+      && ((TARGET_ZBS && popcount_hwi (t) == 32)
+	  || popcount_hwi (t) == 16
+	  || popcount_hwi (t) <= 11))
+    {
+      /* First rotate the relevant bits into the low position.  */
+      int count = BITS_PER_WORD - clz_hwi (~t);
+      rtx x = gen_rtx_ROTATERT (word_mode, operands[1], GEN_INT (count));
+      output = gen_reg_rtx (word_mode);
+      emit_insn (gen_rtx_SET (output, x));
+      input = output;
+
+      /* Now mask off the bits we wanted to clear.  */
+      if (popcount_hwi (t) == 32)
+	{
+	  x = gen_rtx_ZERO_EXTEND (word_mode, gen_lowpart (SImode, input));
+	  emit_insn (gen_rtx_SET (output, x));
+	  input = output;
+	}
+      else if (popcount_hwi (t) == 16)
+	{
+	  x = gen_rtx_ZERO_EXTEND (word_mode, gen_lowpart (HImode, input));
+	  emit_insn (gen_rtx_SET (output, x));
+	  input = output;
+	}
+      else
+	{
+	  rtx t = GEN_INT ((HOST_WIDE_INT_1U << count) - 1);
+	  x = gen_rtx_AND (word_mode, input, t);
+	  emit_insn (gen_rtx_SET (output, x));
+	  input = output;
+	}
+
+      /* And rotate everything back into position.  */
+      x = gen_rtx_ROTATERT (word_mode, input, GEN_INT (BITS_PER_WORD - count));
+      output = gen_reg_rtx (word_mode);
+      emit_insn (gen_rtx_SET (operands[0], x));
+      return true;
+    }
+
   /* The special cases didn't apply.  It's entirely possible we may
      want to combine some of the ideas above with bclr, but for now
      those are deferred until we see them popping up in practice.  */
