diff --git a/gcc/ext-dce.cc b/gcc/ext-dce.cc
index e257e3bc873..5866cc0a3b8 100644
--- a/gcc/ext-dce.cc
+++ b/gcc/ext-dce.cc
@@ -1092,13 +1092,7 @@ ext_dce_rd_transfer_n (int bb_index)
   /* We may have narrowed the set of live objects at the start
      of this block.  If so, update the bitmaps and indicate to
      the generic dataflow code that something changed.  */
-  if (!bitmap_equal_p (&livein[bb_index], livenow))
-    {
-      bitmap_copy (&livein[bb_index], livenow);
-      return true;
-    }
-
-  return false;
+  return bitmap_ior_into (&livein[bb_index], livenow);
 }
 
 /* Dummy function for the df_simple_dataflow API.  */
