diff --git a/gcc/config/riscv/riscv_cmo.h b/gcc/config/riscv/riscv_cmo.h
index 3514fd3f0fe..b331409acb7 100644
--- a/gcc/config/riscv/riscv_cmo.h
+++ b/gcc/config/riscv/riscv_cmo.h
@@ -52,8 +52,6 @@ __riscv_cmo_inval (void *addr)
 
 #if defined (__riscv_zicbop)
 
-# define rnum 1
-
 extern __inline void
 __attribute__ ((__gnu_inline__, __always_inline__, __artificial__))
 __riscv_cmo_prefetch (void *addr, const int vs1, const int vs2)
@@ -65,7 +63,7 @@ extern __inline int
 __attribute__ ((__gnu_inline__, __always_inline__, __artificial__))
 __riscv_cmo_prefetchi ()
 {
-    return __builtin_riscv_zicbop_cbo_prefetchi (rnum);
+    return __builtin_riscv_zicbop_cbo_prefetchi (1);
 }
 
 #endif // __riscv_zicbop
