diff --git a/gcc/match.pd b/gcc/match.pd
index a7b658e9e940..efb8ee838cf1 100644
--- a/gcc/match.pd
+++ b/gcc/match.pd
@@ -9419,6 +9419,7 @@ DEFINE_INT_AND_FLOAT_ROUND_FN (RINT)
  (simplify
   (convert (copysigns (convert@2 @0) (convert @1)))
    (if (optimize
+       && FLOAT_TYPE_P (type)
        && !HONOR_SNANS (@2)
        && types_match (type, TREE_TYPE (@0))
        && types_match (type, TREE_TYPE (@1))
@@ -9430,6 +9431,7 @@ DEFINE_INT_AND_FLOAT_ROUND_FN (RINT)
  (simplify
   (convert (copysigns (convert@2 @0) REAL_CST@1))
    (if (optimize
+       && FLOAT_TYPE_P (type)
        && !HONOR_SNANS (@2)
        && types_match (type, TREE_TYPE (@0))
        && TYPE_PRECISION (type) < TYPE_PRECISION (TREE_TYPE (@2))
@@ -9443,6 +9445,7 @@ DEFINE_INT_AND_FLOAT_ROUND_FN (RINT)
 (simplify
  (convert (abs (convert@1 @0)))
   (if (optimize
+      && FLOAT_TYPE_P (type)
       && !HONOR_SNANS (@1)
       && types_match (type, TREE_TYPE (@0))
       && TYPE_PRECISION (type) < TYPE_PRECISION (TREE_TYPE (@1)))
