diff --git a/gcc/config/riscv/riscv-bclr-lowest-set-bit.cc b/gcc/config/riscv/riscv-bclr-lowest-set-bit.cc
index ecfa511a882b..1dc9ab5c8520 100644
--- a/gcc/config/riscv/riscv-bclr-lowest-set-bit.cc
+++ b/gcc/config/riscv/riscv-bclr-lowest-set-bit.cc
@@ -138,8 +138,8 @@ find_later_ctz (rtx_insn *start, rtx src, int limit)
 		 we need to verify its input doesn't change between
 		 START and NEXT.  We also have to verify that its
 		 destination is unused between those points.  */
-	      if (reg_set_between_p (XEXP (SET_SRC (set), 0), start, next)
-		  || reg_used_between_p (SET_DEST (set), start, next))
+	      if (reg_set_between_p (XEXP (SET_SRC (set), 0), PREV_INSN (start), next)
+		  || reg_used_between_p (SET_DEST (set), PREV_INSN (start), next))
 		return NULL;
 
 	      return next;
@@ -275,7 +275,7 @@ pass_bclr_lowest_set_bit::execute (function *fn)
 	  if (later_ctz)
 	    {
 	      /* Remove the CTZ from the stream and reemit it immediately
-		 after NEXT.  XXX FIXME.  Need to prove this is safe.  */
+		 after NEXT.  */
 	      df_insn_delete (later_ctz);
 	      remove_insn (later_ctz);
 	      SET_PREV_INSN (later_ctz) = NULL;
@@ -290,6 +290,12 @@ pass_bclr_lowest_set_bit::execute (function *fn)
 	      pat = gen_rtx_AND (GET_MODE (dec_dest), pat, dec_src);
 	      pat = gen_rtx_SET (and_dest, pat);
 	      df_insn_rescan (emit_insn_after (pat, NEXT_INSN (next)));
+
+	      /* And remove the original AND.  */
+	      df_insn_delete (next);
+	      remove_insn (next);
+	      SET_PREV_INSN (next) = NULL;
+	      SET_NEXT_INSN (next) = NULL;
 	    }
 	}
     }
diff --git a/gcc/testsuite/gcc.target/riscv/pr127306.c b/gcc/testsuite/gcc.target/riscv/pr127306.c
new file mode 100644
index 000000000000..1c8e5e57ee38
--- /dev/null
+++ b/gcc/testsuite/gcc.target/riscv/pr127306.c
@@ -0,0 +1,30 @@
+/* { dg-do run } */
+/* { dg-additional-options "-march=rv64gc_zbb_zbs -mabi=lp64d" { target rv64 } } */ 
+/* { dg-additional-options "-march=rv32gc_zbb_zbs -mabi=ilp32" { target rv32 } } */ 
+
+typedef unsigned short uint16_t;
+
+__attribute__((noinline)) unsigned
+scan(uint16_t mask)
+{
+    unsigned result = 0, dword = mask, bit = 0;
+
+    while (dword) {
+        result |= 1u << bit;
+	dword &= dword - 1;
+	bit = __builtin_ffs(dword) - 1;
+    }
+
+    return result;
+}
+
+int
+main(void)
+{
+    const unsigned expected = 0xffff;
+    const unsigned actual = scan(expected);
+
+    if (actual != expected)
+      __builtin_abort ();
+    __builtin_exit (0);
+}
