diff --git a/gcc/ext-dce.cc b/gcc/ext-dce.cc
index 626c431f601..b2b92d74761 100644
--- a/gcc/ext-dce.cc
+++ b/gcc/ext-dce.cc
@@ -893,15 +893,20 @@ ext_dce_process_bb (basic_block bb)
 {
   rtx_insn *insn;
 
+  /* Live-out state of the destination of this insn.  We can
+     use this to refine the live-in state of the sources of
+     this insn in many cases.  */
+  bitmap live_tmp = BITMAP_ALLOC (NULL);
+
   FOR_BB_INSNS_REVERSE (bb, insn)
     {
       if (!NONDEBUG_INSN_P (insn))
 	continue;
 
-      /* Live-out state of the destination of this insn.  We can
-	 use this to refine the live-in state of the sources of
-	 this insn in many cases.  */
-      bitmap live_tmp = BITMAP_ALLOC (NULL);
+      /* We only allocate/deallocate the bitmap once and reuse it
+	 within this loop.  So clear it to make sure it is empty for
+	 each insn pocessed.  */
+      bitmap_clear (live_tmp);
 
       /* First process any sets/clobbers in INSN.  */
       bool skipped_dest = ext_dce_process_sets (insn, PATTERN (insn), live_tmp);
@@ -939,8 +944,9 @@ ext_dce_process_bb (basic_block bb)
 	  ext_dce_process_uses (insn, CALL_INSN_FUNCTION_USAGE (insn), live_tmp, false);
 	}
 
-      BITMAP_FREE (live_tmp);
     }
+
+  BITMAP_FREE (live_tmp);
 }
 
 /* SUBREG_PROMOTED_VAR_P is set by the gimple->rtl optimizers and
