diff --git a/gcc/config/mcore/mcore.h b/gcc/config/mcore/mcore.h
index 3de4f0b336a..0a396e8a496 100644
--- a/gcc/config/mcore/mcore.h
+++ b/gcc/config/mcore/mcore.h
@@ -82,7 +82,7 @@
 extern char * mcore_current_function_name;
  
 /* The MCore ABI says that bitfields are unsigned by default.  */
-#define CC1_SPEC "-funsigned-bitfields"
+//#define CC1_SPEC "-funsigned-bitfields"
 
 /* Target machine storage Layout.  */
 
