diff --git a/gcc/config/riscv/riscv.cc b/gcc/config/riscv/riscv.cc
index 439cc12f93d..88a85932410 100644
--- a/gcc/config/riscv/riscv.cc
+++ b/gcc/config/riscv/riscv.cc
@@ -5250,13 +5250,18 @@ riscv_expand_conditional_branch (rtx label, rtx_code code, rtx op0, rtx op1)
    Return 0 if expansion failed.  */
 
 bool
-riscv_expand_conditional_move (rtx dest, rtx op, rtx cons, rtx alt)
+riscv_expand_conditional_move (rtx dest_, rtx op, rtx cons, rtx alt)
 {
-  machine_mode mode = GET_MODE (dest);
+  machine_mode mode = GET_MODE (dest_);
   rtx_code code = GET_CODE (op);
   rtx op0 = XEXP (op, 0);
   rtx op1 = XEXP (op, 1);
 
+  /* The destination must be in WORD_MODE.  But we can use the trick of
+     storing into an suitable pseudo, then using a lowpart extraction.  */
+  rtx dest = (mode != word_mode) ? gen_reg_rtx (word_mode) : dest_;
+  mode = word_mode;
+
   if (((TARGET_ZICOND_LIKE
 	|| (arith_operand (cons, mode) && arith_operand (alt, mode)))
        && (GET_MODE_CLASS (mode) == MODE_INT))
@@ -5265,13 +5270,16 @@ riscv_expand_conditional_move (rtx dest, rtx op, rtx cons, rtx alt)
       machine_mode mode0 = GET_MODE (op0);
       machine_mode mode1 = GET_MODE (op1);
 
-      /* An integer comparison must be comparing WORD_MODE objects.  We
-	 must enforce that so that we don't strip away a sign_extension
-	 thinking it is unnecessary.  We might consider using
-	 riscv_extend_operands if they are not already properly extended.  */
+      /* An integer comparison must be comparing WORD_MODE objects.
+	 Extend the comparison arguments as necessary.  */
       if ((INTEGRAL_MODE_P (mode0) && mode0 != word_mode)
 	  || (INTEGRAL_MODE_P (mode1) && mode1 != word_mode))
-	return false;
+	riscv_extend_comparands (code, &op0, &op1);
+
+      if (!REG_P (op0) && !CONST_INT_P (op0))
+	op0 = force_reg (mode, op0);
+      if (!REG_P (op1) && !CONST_INT_P (op1))
+	op1 = force_reg (mode, op1);
 
       /* In the fallback generic case use MODE rather than WORD_MODE for
 	 the output of the SCC instruction, to match the mode of the NEG
@@ -5318,6 +5326,21 @@ riscv_expand_conditional_move (rtx dest, rtx op, rtx cons, rtx alt)
       else if (!TARGET_ZICOND_LIKE && !TARGET_SFB_ALU && !TARGET_XTHEADCONDMOV)
 	riscv_expand_int_scc (tmp, code, op0, op1, &invert);
 
+      /* Precondition the true/false arms of the IF_THEN_ELSE by sign extending
+	 any sub-word operand to WORD_MODE.  */
+      if (GET_MODE (cons) != VOIDmode && GET_MODE (cons) != word_mode)
+	{
+	  rtx new_rtx = gen_reg_rtx (word_mode);
+	  convert_move (new_rtx, cons, 0);
+	  cons = new_rtx;
+	}
+      if (GET_MODE (alt) != VOIDmode && GET_MODE (alt) != word_mode)
+	{
+	  rtx new_rtx = gen_reg_rtx (word_mode);
+	  convert_move (new_rtx, alt, 0);
+	  alt = new_rtx;
+	}
+
       if (TARGET_SFB_ALU || TARGET_XTHEADCONDMOV)
 	{
 	  riscv_emit_int_compare (&code, &op0, &op1, !TARGET_SFB_ALU);
@@ -5332,6 +5355,7 @@ riscv_expand_conditional_move (rtx dest, rtx op, rtx cons, rtx alt)
 	  alt = force_reg (mode, alt);
 	  emit_insn (gen_rtx_SET (dest, gen_rtx_IF_THEN_ELSE (mode, cond,
 							      cons, alt)));
+	  goto success;
 	  return true;
 	}
       else if (!TARGET_ZICOND_LIKE)
@@ -5349,7 +5373,7 @@ riscv_expand_conditional_move (rtx dest, rtx op, rtx cons, rtx alt)
 	  riscv_emit_unary (NOT, reg3, reg1);
 	  riscv_emit_binary (AND, reg4, reg3, alt);
 	  riscv_emit_binary (IOR, dest, reg2, reg4);
-	  return true;
+	  goto success;
 	}
       /* 0, reg or 0, imm */
       else if (cons == CONST0_RTX (mode)
@@ -5362,7 +5386,7 @@ riscv_expand_conditional_move (rtx dest, rtx op, rtx cons, rtx alt)
 	  emit_insn (gen_rtx_SET (dest,
 				  gen_rtx_IF_THEN_ELSE (mode, cond,
 							cons, alt)));
-	  return true;
+	  goto success;
 	}
       /* imm, imm */
       else if (CONST_INT_P (cons) && cons != CONST0_RTX (mode)
@@ -5382,7 +5406,7 @@ riscv_expand_conditional_move (rtx dest, rtx op, rtx cons, rtx alt)
 	  if (!SMALL_OPERAND (INTVAL (cons)))
 	    cons = force_reg (mode, cons);
 	  riscv_emit_binary (PLUS, dest, dest, cons);
-	  return true;
+	  goto success;
 	}
       /* imm, reg  */
       else if (CONST_INT_P (cons) && cons != CONST0_RTX (mode) && REG_P (alt))
@@ -5395,7 +5419,7 @@ riscv_expand_conditional_move (rtx dest, rtx op, rtx cons, rtx alt)
 	      emit_insn (gen_rtx_SET (dest,
 				      gen_rtx_IF_THEN_ELSE (mode, cond,
 							    cons, alt)));
-	      return true;
+	      goto success;
 	    }
 
 	  riscv_emit_int_compare (&code, &op0, &op1, true);
@@ -5418,19 +5442,27 @@ riscv_expand_conditional_move (rtx dest, rtx op, rtx cons, rtx alt)
 							CONST0_RTX (mode),
 							temp1)));
 	  riscv_emit_binary (PLUS, dest, dest, cons);
-	  return true;
+	  goto success;
 	}
       /* reg, 0 or imm, 0  */
       else if ((REG_P (cons)
+#if 0
+		|| (SUBREG_P (cons)
+		    && SUBREG_PROMOTED_VAR_P (cons)
+		    && SUBREG_PROMOTED_SIGNED_P (cons))
+#endif
 		|| (CONST_INT_P (cons) && cons != CONST0_RTX (mode)))
 	       && alt == CONST0_RTX (mode))
 	{
+	  if (SUBREG_P (cons))
+	    cons = SUBREG_REG (cons);
+
 	  riscv_emit_int_compare (&code, &op0, &op1, true);
 	  rtx cond = gen_rtx_fmt_ee (code, GET_MODE (op0), op0, op1);
 	  cons = force_reg (mode, cons);
 	  emit_insn (gen_rtx_SET (dest, gen_rtx_IF_THEN_ELSE (mode, cond,
 							      cons, alt)));
-	  return true;
+	  goto success;
 	}
       /* reg, imm  */
       else if (REG_P (cons) && CONST_INT_P (alt) && alt != CONST0_RTX (mode))
@@ -5443,7 +5475,7 @@ riscv_expand_conditional_move (rtx dest, rtx op, rtx cons, rtx alt)
 	      emit_insn (gen_rtx_SET (dest,
 				      gen_rtx_IF_THEN_ELSE (mode, cond,
 							    cons, alt)));
-	      return true;
+	      goto success;
 	    }
 
 	  riscv_emit_int_compare (&code, &op0, &op1, true);
@@ -5466,7 +5498,7 @@ riscv_expand_conditional_move (rtx dest, rtx op, rtx cons, rtx alt)
 							temp1,
 							CONST0_RTX (mode))));
 	  riscv_emit_binary (PLUS, dest, dest, alt);
-	  return true;
+	  goto success;
 	}
       /* reg, reg  */
       else if (REG_P (cons) && REG_P (alt))
@@ -5480,7 +5512,7 @@ riscv_expand_conditional_move (rtx dest, rtx op, rtx cons, rtx alt)
 	      emit_insn (gen_rtx_SET (dest,
 				      gen_rtx_IF_THEN_ELSE (mode, cond,
 							    cons, alt)));
-	      return true;
+	      goto success;
 	    }
 
 	  rtx reg1 = gen_reg_rtx (mode);
@@ -5498,11 +5530,21 @@ riscv_expand_conditional_move (rtx dest, rtx op, rtx cons, rtx alt)
 							CONST0_RTX (mode),
 							alt)));
 	  riscv_emit_binary (PLUS, dest, reg1, reg2);
-	  return true;
+	  goto success;
 	}
     }
 
   return false;
+
+ success:
+  if (dest != dest_)
+    {
+      dest = gen_lowpart (GET_MODE (dest_), dest);
+      SUBREG_PROMOTED_VAR_P (dest) = 1;
+      SUBREG_PROMOTED_SET (dest, SRP_SIGNED);
+      emit_move_insn (dest_, dest);
+    }
+  return true;
 }
 
 /* Implement TARGET_FUNCTION_ARG_BOUNDARY.  Every parameter gets at
diff --git a/gcc/testsuite/gcc.target/riscv/pr111126.c b/gcc/testsuite/gcc.target/riscv/pr111126.c
new file mode 100644
index 00000000000..f305c31eb4d
--- /dev/null
+++ b/gcc/testsuite/gcc.target/riscv/pr111126.c
@@ -0,0 +1,36 @@
+/* { dg-do compile } */
+/* { dg-options "-march=rv64gc_zicond -mabi=lp64d" { target rv64 } } 
+/* { dg-options "-march=rv32gc_zicond -mabi=ilp32" { target rv32 } } 
+/* { dg-skip-if "" { *-*-* } { "-O0" "-O1" "-Os" "-Oz" "-O3" "-Og" } } */
+
+int f(bool a, int c)
+{
+  return a * c;
+}
+
+int f0(bool a, int c)
+{
+  return (-a) & c;
+}
+
+int f1(bool a, int c)
+{
+  return a ? c : 0;
+}
+
+int f3(int a, int b, int c)
+{
+  return (a == b) * c;
+}
+int g0(bool a, int c)
+{
+  return a ? 0 : c;
+}
+int g1(bool a, int c)
+{
+  a = !a;
+  return a * c;
+}
+
+/* { dg-final { scan-assembler-times "czero" 6 } } */
+
diff --git a/gcc/testsuite/gcc.target/riscv/pr118384.c b/gcc/testsuite/gcc.target/riscv/pr118384.c
new file mode 100644
index 00000000000..dae65a3d28c
--- /dev/null
+++ b/gcc/testsuite/gcc.target/riscv/pr118384.c
@@ -0,0 +1,25 @@
+/* { dg-do compile } */
+/* { dg-options "-O2 -march=rv64gc_zicond -mabi=lp64d" { target rv64 } } */
+/* { dg-options "-O2 -march=rv32gc_zicond -mabi=ilp32" { target rv32 } } */
+/* { dg-skip-if "" { *-*-* } { "-O0" "-O1" "-Os" "-Oz" "-O3" "-Og" } } */
+
+#include <stdint-gcc.h>
+uint64_t muldi3(uint64_t a, uint64_t b)
+{
+
+    uint64_t Product;
+
+    Product = 0;
+    while (a) {
+        if (a & 1) {
+            Product += b;
+        }
+        a >>= 1;
+        b <<= 1;
+    }
+
+    return Product;
+}
+
+/* { dg-final { scan-assembler-times "czero" 1 } } */
+
