diff --git a/gcc/testsuite/gcc.target/arc/pic-2.c b/gcc/testsuite/gcc.target/arc/pic-2.c
index 4b0e17126e91..e9997b34f3de 100644
--- a/gcc/testsuite/gcc.target/arc/pic-2.c
+++ b/gcc/testsuite/gcc.target/arc/pic-2.c
@@ -1,6 +1,6 @@
 /* { dg-do compile } */
 /* { dg-skip-if "PIC not available for ARC6xx" { arc6xx } } */
-/* { dg-options "-mno-sdata -O2 -fpic -fno-builtin" } */
+/* { dg-options "-mno-sdata -O2 -fpic -fno-builtin -fno-tree-loop-ivcanon" } */
 
 /* Check if we resolve correctly complex PIC addresses.  */
 
