commit 28132c5bbd86887afc7b41175ed3d0dfe579d750
Author: Jeff Law <jeffrey.law@oss.qualcomm.com>
Date:   Tue Mar 17 11:01:23 2026 -0600

    Daniel's patch for pr71336

diff --git a/gcc/match.pd b/gcc/match.pd
index 1d6428bf7e53..73bf697997aa 100644
--- a/gcc/match.pd
+++ b/gcc/match.pd
@@ -5195,6 +5195,44 @@ DEFINE_INT_AND_FLOAT_ROUND_FN (RINT)
        && expr_no_side_effects_p (@2))
        (op (mult (convert:type @0) @2) @1))))
 
+/* PR71336:
+   zero_one != 0 ? CST1 : CST2 -> ((typeof (CST2))zero_one * diff) + CST2,
+   where CST1 > CST2 and diff = CST1 - CST2.
+
+   Includes the "zero_one == 0 ? (...)" variant too.  */
+(for cmp (ne eq)
+ (simplify
+  (cond (cmp zero_one_valued_p@0 integer_zerop) INTEGER_CST@1 INTEGER_CST@2)
+  (with {
+    unsigned HOST_WIDE_INT diff = 0;
+
+    if (tree_int_cst_sgn (@1) > 0 && tree_int_cst_sgn (@2) > 0
+	&& tree_fits_uhwi_p (@1) && tree_fits_uhwi_p (@2))
+     {
+	if (cmp == NE_EXPR
+	    && wi::gtu_p (wi::to_wide (@1), wi::to_wide (@2)))
+	  diff = tree_to_uhwi (@1) - tree_to_uhwi (@2);
+
+	if (cmp == EQ_EXPR
+	    && wi::gtu_p (wi::to_wide (@2), wi::to_wide (@1)))
+	  diff = tree_to_uhwi (@2) - tree_to_uhwi (@1);
+     }
+   }
+   (if (cmp == NE_EXPR
+	&& 0 && INTEGRAL_TYPE_P (type)
+	&& TYPE_PRECISION (type) <= BITS_PER_WORD
+	&& INTEGRAL_TYPE_P (TREE_TYPE (@0))
+	&& diff > 0)
+     (plus (mult (convert:type @0) { build_int_cst (type, diff); })
+	    @2)
+    (if (cmp == EQ_EXPR
+	 && 0 && INTEGRAL_TYPE_P (type)
+	 && TYPE_PRECISION (type) <= BITS_PER_WORD
+	 && INTEGRAL_TYPE_P (TREE_TYPE (@0))
+	 && diff > 0)
+      (plus (mult (convert:type @0) { build_int_cst (type, diff); })
+	     @1))))))
+
 /* ?: Value replacement. */
 /* a == 0 ? b : b + a  -> b + a */
 (for op (plus bit_ior bit_xor)
