diff --git a/libgcc/strub.c b/libgcc/strub.c
index b0f990d9dee..346351cd492 100644
--- a/libgcc/strub.c
+++ b/libgcc/strub.c
@@ -138,12 +138,12 @@ __strub_leave (void **mark)
     *base++ = 0;
   while (base < end);
   /* Make sure the stack overwrites are not optimized away.  */
-  asm ("" : : "m" (end[0]));
+  /* asm ("" : : "m" (end[0])); */
 #else
   do
     *--end = 0;
   while (base < end);
   /* Make sure the stack overwrites are not optimized away.  */
-  asm ("" : : "m" (base[0]));
+  /* asm ("" : : "m" (base[0])); */
 #endif
 }
