diff --git a/gcc/testsuite/gcc.c-torture/compile/attr-complex-method-2.c b/gcc/testsuite/gcc.c-torture/compile/attr-complex-method-2.c
index dc28e2c99c6..19ec1dbdf0c 100644
--- a/gcc/testsuite/gcc.c-torture/compile/attr-complex-method-2.c
+++ b/gcc/testsuite/gcc.c-torture/compile/attr-complex-method-2.c
@@ -8,5 +8,5 @@ void do_div (_Complex double *a, _Complex double *b)
   *a = *b / (4.0 - 5.0fi);
 }
 
-/* { dg-final { scan-tree-dump "__(?:gnu_)?divdc3" "optimized" { target { ! { avr-*-* } } } } } */
-/* { dg-final { scan-tree-dump "__(?:gnu_)?divsc3" "optimized" { target { avr-*-* } } } } */
+/* { dg-final { scan-tree-dump "__(?:gnu_)?divdc3" "optimized" { target { large_double } } } } */
+/* { dg-final { scan-tree-dump "__(?:gnu_)?divsc3" "optimized" { target { ! { large_double } } } } } */
diff --git a/gcc/testsuite/gcc.c-torture/compile/attr-complex-method.c b/gcc/testsuite/gcc.c-torture/compile/attr-complex-method.c
index 046de7efeb9..239f2f7d36f 100644
--- a/gcc/testsuite/gcc.c-torture/compile/attr-complex-method.c
+++ b/gcc/testsuite/gcc.c-torture/compile/attr-complex-method.c
@@ -8,4 +8,6 @@ void do_div (_Complex double *a, _Complex double *b)
   *a = *b / (4.0 - 5.0fi);
 }
 
-/* { dg-final { scan-tree-dump-not "__divdc3" "optimized" } } */
+/* { dg-final { scan-tree-dump-not "__divdc3" "optimized" { target { large_double } } } } */
+/* { dg-final { scan-tree-dump-not "__divsc3" "optimized" { target { ! { large_double } } } } } */
+
