commit bc6dcdf87c5c3b0d9352e5a40ccd322df46f5418
Author: Jeff Law <jeffrey.law@oss.qualcomm.com>
Date:   Sun Mar 15 23:09:48 2026 -0600

    Do not consider relaxed atomic loads inherently necessary.

diff --git a/gcc/tree-ssa-dce.cc b/gcc/tree-ssa-dce.cc
index 7932607e889..7d31875c224 100644
--- a/gcc/tree-ssa-dce.cc
+++ b/gcc/tree-ssa-dce.cc
@@ -48,6 +48,7 @@ along with GCC; see the file COPYING3.  If not see
 #include "backend.h"
 #include "rtl.h"
 #include "tree.h"
+#include "memmodel.h"
 #include "gimple.h"
 #include "cfghooks.h"
 #include "tree-pass.h"
@@ -419,6 +420,27 @@ mark_stmt_if_obviously_necessary (gimple *stmt, bool aggressive)
 	    mark_stmt_necessary (call, true);
 	    return;
 	  }
+
+	if (gimple_call_builtin_p (stmt, BUILT_IN_NORMAL))
+	  {
+	    tree callee = gimple_call_fndecl (stmt);
+	    switch (DECL_FUNCTION_CODE (callee))
+	      {
+	      case BUILT_IN_ATOMIC_LOAD_1:
+	      case BUILT_IN_ATOMIC_LOAD_2:
+	      case BUILT_IN_ATOMIC_LOAD_4:
+	      case BUILT_IN_ATOMIC_LOAD_8:
+	      case BUILT_IN_ATOMIC_LOAD_16:
+		if (TREE_CODE (gimple_call_arg (stmt, 1)) == INTEGER_CST
+		    && TREE_INT_CST_LOW (gimple_call_arg (stmt, 1)) == MEMMODEL_RELAXED)
+		  return;
+
+	      default:
+		break;
+	      }
+
+	  }
+
 	break;
       }
 
