diff --git a/gcc/testsuite/gcc.target/riscv/pr124766.c b/gcc/testsuite/gcc.target/riscv/pr124766.c
new file mode 100644
index 000000000000..b16c31e8a9bd
--- /dev/null
+++ b/gcc/testsuite/gcc.target/riscv/pr124766.c
@@ -0,0 +1,17 @@
+/* { dg-do compile } */
+/* { dg-options "-march=rv64gcbv_zicond -mabi=lp64d" { target rv64 } } */
+/* { dg-options "-march=rv32gcbv_zicond -mabi=ilp32" { target rv32 } } */
+/* { dg-skip-if "" { *-*-* } { "-O0" "-Og" } } */
+
+long
+frob (unsigned short *y, long z)
+{
+  long ret = (*y << 2) + z;
+  if (ret != z)
+    return 0;
+  return ret;
+}
+
+/* { dg-final { scan-assembler-not "sh2add" } } */
+/* { dg-final { scan-assembler-not "sub" } } */
+
diff --git a/gcc/simplify-rtx.cc b/gcc/simplify-rtx.cc
index 50fc51152cab..23de16800065 100644
--- a/gcc/simplify-rtx.cc
+++ b/gcc/simplify-rtx.cc
@@ -6649,6 +6649,14 @@ simplify_context::simplify_relational_operation_1 (rtx_code code,
       return simplify_gen_relational (code, mode, cmp_mode, x, tem);
     }
 
+  /* (eq/ne (plus (x) (y)) y) simplifies to (eq/ne x 0).  */
+  if ((code == EQ || code == NE)
+      && op0code == PLUS
+      && XEXP (op0, 1) == op1
+      && (INTEGRAL_MODE_P (cmp_mode) || flag_unsafe_math_optimizations))
+    return simplify_gen_relational (code, mode, cmp_mode,
+				    XEXP (op0, 0), const0_rtx);
+
   /* (ne:SI (zero_extract:SI FOO (const_int 1) BAR) (const_int 0))) is
      the same as (zero_extract:SI FOO (const_int 1) BAR).  */
   scalar_int_mode int_mode, int_cmp_mode;
