
diff --git a/gcc/config/riscv/riscv.md b/gcc/config/riscv/riscv.md
index 3fe0ad0ccdf4..1cc0c6c72d47 100644
--- a/gcc/config/riscv/riscv.md
+++ b/gcc/config/riscv/riscv.md
@@ -4960,6 +4960,48 @@ (define_split
   { operands[3] = GEN_INT (BITS_PER_WORD
 			   - exact_log2 (INTVAL (operands[3]) + 1)); })
 
+;; EQ/NE of a sign bit splat against zero is just GE/LT 0, so we can
+;; recognize it directly.  Note there may be a subreg expression buried
+;; in there
+(define_insn "*sign_bit_splat_equality_test"
+  [(set (pc)
+	(if_then_else
+	 (any_eq
+	  (subreg:SI (ashiftrt:DI (match_operand:DI 1 "register_operand" "r")
+				  (const_int 63)) 0)
+	  (const_int 0))
+	 (label_ref (match_operand 0 "" ""))
+	 (pc)))]
+  "TARGET_64BIT"
+{
+  rtx x = PATTERN (insn);
+
+  /* We'll always have a SET, so it's safe to extract the source.  */
+  x = SET_SRC (x);
+
+  /* Get the condition of the IF_THEN_ELSE.  */
+  x = XEXP (x, 0);
+
+  if (GET_CODE (x) == EQ)
+    {
+      if (get_attr_length (insn) == 12)
+	return "blt\t%1,zero,1f; jump\t%l0,ra; 1:";
+      return "bge\t%1,zero,%l0";
+    }
+  else if (GET_CODE (x) == NE)
+    {
+      if (get_attr_length (insn) == 12)
+	return "bge\t%1,zero,1f; jump\t%l0,ra; 1:";
+      return "blt\t%1,zero,%l0";
+    }
+  else
+    gcc_unreachable ();
+}
+  [(set_attr "type" "branch")
+   (set_attr "mode" "none")])
+					
+	
+
 ;; Standard extensions and pattern for optimization
 (include "bitmanip.md")
 (include "crypto.md")
diff --git a/gcc/testsuite/gcc.target/riscv/sign-extend-1.c b/gcc/testsuite/gcc.target/riscv/sign-extend-1.c
index d8c18dd1aaa7..bba653321cc8 100644
--- a/gcc/testsuite/gcc.target/riscv/sign-extend-1.c
+++ b/gcc/testsuite/gcc.target/riscv/sign-extend-1.c
@@ -7,7 +7,9 @@ foo1 (int i)
 {
   return i >> 31;
 }
-/* { dg-final { scan-assembler "sraiw\ta\[0-9\],a\[0-9\],31" } } */
+/* simplify-rtx recognizes this as a sign bit splat which it may promote
+   to 64 bits wide, so srai, not sraiw.  */
+/* { dg-final { scan-assembler "srai\ta\[0-9\],a\[0-9\],63" } } */
 
 signed char
 sub2 (long i)
@@ -23,6 +25,5 @@ sub3 (long i)
 }
 /* { dg-final { scan-assembler "sraiw\ta\[0-9\],a\[0-9\],16" } } */
 
-/* { dg-final { scan-assembler-not "srai\t" } } */
 /* { dg-final { scan-assembler-not "srli\t" } } */
 /* { dg-final { scan-assembler-not "srliw\t" } } */
diff --git a/gcc/testsuite/gcc.target/riscv/sign-extend-rshift.c b/gcc/testsuite/gcc.target/riscv/sign-extend-rshift.c
index bcb015efe03d..5512fc738160 100644
--- a/gcc/testsuite/gcc.target/riscv/sign-extend-rshift.c
+++ b/gcc/testsuite/gcc.target/riscv/sign-extend-rshift.c
@@ -118,6 +118,8 @@ SLONG_EXT_SSHORT_RSHIFT_N_SINT(15)
 /* { dg-final { scan-assembler-times "slli\t" 17 { target { rv32 } } } } */
 /* { dg-final { scan-assembler-times "srai\t" 26 { target { rv32 } } } } */
 
+/* One of the natural sraiw cases is a 32 bit sign bit splat which gets
+   promoted to 64 bits wide.  */
 /* { dg-final { scan-assembler-times "slli\t" 44 { target { rv64 } } } } */
-/* { dg-final { scan-assembler-times "srai\t" 58 { target { rv64 } } } } */
-/* { dg-final { scan-assembler-times "sraiw\t" 3 { target { rv64 } } } } */
+/* { dg-final { scan-assembler-times "srai\t" 59 { target { rv64 } } } } */
+/* { dg-final { scan-assembler-times "sraiw\t" 2 { target { rv64 } } } } */
