diff --git a/gcc/simplify-rtx.cc b/gcc/simplify-rtx.cc
index 50fc51152cab..0b59029fd368 100644
--- a/gcc/simplify-rtx.cc
+++ b/gcc/simplify-rtx.cc
@@ -6649,6 +6739,35 @@ simplify_context::simplify_relational_operation_1 (rtx_code code,
       return simplify_gen_relational (code, mode, cmp_mode, x, tem);
     }
 
+  /* (eq/ne (plus (x) (y)) y) simplifies to (eq/ne x 0).  */
+  if ((code == EQ || code == NE)
+      && op0code == PLUS
+      && XEXP (op0, 1) == op1
+      && (INTEGRAL_MODE_P (cmp_mode) || flag_unsafe_math_optimizations))
+    return simplify_gen_relational (code, mode, cmp_mode,
+				    XEXP (op0, 0), const0_rtx);
+
+  /* (eq/ne (sign-bit-splat) (const_int 0)) is just (ge 0) and (lt 0) respectively.
+     We may have a SUBREG wrapping the sign-bit splat.  */
+  if ((code == EQ || code == NE)
+      && op1 == const0_rtx)
+    {
+      rtx new_op0 = op0;
+
+      if (GET_CODE (op0) == SUBREG && subreg_lowpart_p (op0))
+	new_op0 = SUBREG_REG (op0);
+
+      if (GET_CODE (new_op0) == ASHIFTRT
+	  && CONST_INT_P (XEXP (new_op0, 1))
+	  && known_eq (GET_MODE_BITSIZE (GET_MODE (new_op0)),
+					 INTVAL (XEXP (new_op0, 1)) + 1))
+	return simplify_gen_relational (code == EQ ? GE : LT,
+					mode,
+					GET_MODE (new_op0),
+					XEXP (new_op0, 0),
+					const0_rtx);
+    }
+
   /* (ne:SI (zero_extract:SI FOO (const_int 1) BAR) (const_int 0))) is
      the same as (zero_extract:SI FOO (const_int 1) BAR).  */
   scalar_int_mode int_mode, int_cmp_mode;
