diff --git a/ld/scripttempl/v850.sc b/ld/scripttempl/v850.sc
index ddf5a44385a..b8d8dba1290 100644
--- a/ld/scripttempl/v850.sc
+++ b/ld/scripttempl/v850.sc
@@ -211,7 +211,7 @@ EOF
 
 cat <<EOF
   /* User stack.  */
-  .stack 0x200000	:
+  .stack 0x300000	:
   {
 	${RELOCATING+__stack = .;}
 	*(.stack)
diff --git a/ld/scripttempl/v850_rh850.sc b/ld/scripttempl/v850_rh850.sc
index fba59cc242b..7dac9976cfc 100644
--- a/ld/scripttempl/v850_rh850.sc
+++ b/ld/scripttempl/v850_rh850.sc
@@ -231,7 +231,7 @@ EOF
 
 cat <<EOF
   /* User stack.  */
-  .stack 0x200000	:
+  .stack 0x300000	:
   {
 	${RELOCATING+__stack = .;}
 	*(.stack)
diff --git a/sim/v850/v850-sim.h b/sim/v850/v850-sim.h
index 35032306235..70b8a680cb9 100644
--- a/sim/v850/v850-sim.h
+++ b/sim/v850/v850-sim.h
@@ -46,7 +46,7 @@ extern SIM_DESC simulator;
 
 
 #define V850_ROM_SIZE 0x8000
-#define V850_LOW_END 0x200000
+#define V850_LOW_END 0x300000
 #define V850_HIGH_START 0xffe000
 
 
