diff --git a/ld/scripttempl/v850.sc b/ld/scripttempl/v850.sc
index ddf5a44385a..b8d8dba1290 100644
--- a/ld/scripttempl/v850.sc
+++ b/ld/scripttempl/v850.sc
@@ -211,7 +211,7 @@ EOF
 
 cat <<EOF
   /* User stack.  */
-  .stack 0x200000	:
+  .stack 0x300000	:
   {
 	${RELOCATING+__stack = .;}
 	*(.stack)
diff --git a/ld/scripttempl/v850_rh850.sc b/ld/scripttempl/v850_rh850.sc
index fba59cc242b..7dac9976cfc 100644
--- a/ld/scripttempl/v850_rh850.sc
+++ b/ld/scripttempl/v850_rh850.sc
@@ -231,7 +231,7 @@ EOF
 
 cat <<EOF
   /* User stack.  */
-  .stack 0x200000	:
+  .stack 0x300000	:
   {
 	${RELOCATING+__stack = .;}
 	*(.stack)
diff --git a/sim/v850/sim-main.h b/sim/v850/sim-main.h
index 49b845dfe7a..26839da44da 100644
--- a/sim/v850/sim-main.h
+++ b/sim/v850/sim-main.h
@@ -48,7 +48,7 @@ extern SIM_DESC simulator;
 
 
 #define V850_ROM_SIZE 0x8000
-#define V850_LOW_END 0x200000
+#define V850_LOW_END 0x300000
 #define V850_HIGH_START 0xffe000
 
 
diff --git a/sim/v850/simops.c b/sim/v850/simops.c
index 8fac8bd9891..94399e2ad4f 100644
--- a/sim/v850/simops.c
+++ b/sim/v850/simops.c
@@ -339,10 +339,10 @@ Multiply64 (int sign, unsigned long op0)
 	  
       sign = (op0 ^ op1) & 0x80000000;
 	  
-      if (((signed long) op0) < 0)
+      if (op0 & 0x80000000)
 	op0 = - op0;
 	  
-      if (((signed long) op1) < 0)
+      if (op1 & 0x80000000)
 	op1 = - op1;
     }
       
@@ -3135,8 +3135,8 @@ v850_div (SIM_DESC sd, unsigned int op0, unsigned int op1, unsigned int *op2p, u
   bfd_boolean     overflow = FALSE;
   
   /* Compute the result.  */
-  divide_by   = op0;
-  divide_this = op1;
+  divide_by   = (int32_t) op0;
+  divide_this = (int32_t) op1;
 
   if (divide_by == 0 || (divide_by == -1 && divide_this == (1 << 31)))
     {
@@ -3267,7 +3267,7 @@ v850_bins (SIM_DESC sd, unsigned int source, unsigned int lsb, unsigned int msb,
   pos = lsb;
   width = (msb - lsb) + 1;
 
-  mask = ~ (-(1 << width));
+  mask = ~ (-(1UL << width));
   source &= mask;
   mask <<= pos;
   result = (* dest) & ~ mask;
