diff --git a/sim/mcore/interp.c b/sim/mcore/interp.c
index 48d9ff8645a..15f15b8563a 100644
--- a/sim/mcore/interp.c
+++ b/sim/mcore/interp.c
@@ -1034,7 +1034,7 @@ step_once (SIM_DESC sd, SIM_CPU *cpu)
 	case 0x3A: case 0x3B:				/* asrc, asri */
 	  {
 	    unsigned imm = IMM5;
-	    long tmp = gr[RD];
+	    uint32_t tmp = gr[RD];
 	    if (imm == 0)
 	      {
 		NEW_C (tmp);
@@ -1060,7 +1060,7 @@ step_once (SIM_DESC sd, SIM_CPU *cpu)
 	case 0x3E: case 0x3F:				/* lsrc, lsri */
 	  {
 	    unsigned imm = IMM5;
-	    unsigned long tmp = gr[RD];
+	    uint32_t tmp = gr[RD];
 	    if (imm == 0)
 	      {
 		NEW_C (tmp);
