diff --git a/bfd/elflink.c b/bfd/elflink.c
index 107480286f3..99542b7a524 100644
--- a/bfd/elflink.c
+++ b/bfd/elflink.c
@@ -11832,6 +11832,7 @@ elf_link_input_bfd (struct elf_final_link_info *flinfo, bfd *input_bfd)
 		if ((o->flags & SEC_ELF_REVERSE_COPY)
 		    && o->size > address_size)
 		  {
+#if 0
 		    /* Reverse-copy input section to output.  */
 
 		    if (o->reloc_count != 0
@@ -11846,6 +11847,7 @@ elf_link_input_bfd (struct elf_final_link_info *flinfo, bfd *input_bfd)
 			bfd_set_error (bfd_error_bad_value);
 			return false;
 		      }
+#endif
 
 		    do
 		      {
