diff --git a/ld/configure.tgt b/ld/configure.tgt
index 66b81550458..3c58ed1ee1c 100644
--- a/ld/configure.tgt
+++ b/ld/configure.tgt
@@ -272,7 +272,11 @@ epiphany-*-*)		targ_emul=elf32epiphany
 fido*-*-elf*)		targ_emul=m68kelf
 			;;
 fr30-*-*)		targ_emul=elf32fr30
-			targ_extra_ofiles=ldelfgen.o
+			# The default linker script yields just one segment
+			# as intended, and then a warning is not helpful.
+			if test "${ac_default_ld_warn_rwx_segments}" = unset; then
+			  ac_default_ld_warn_rwx_segments=0
+			fi
 			;;
 frv-*-*linux*)		targ_emul=elf32frvfd
 			;;
diff --git a/ld/emulparams/elf32fr30.sh b/ld/emulparams/elf32fr30.sh
index 1764a325698..9eec80270a1 100644
--- a/ld/emulparams/elf32fr30.sh
+++ b/ld/emulparams/elf32fr30.sh
@@ -1,7 +1,6 @@
 MACHINE=
 SCRIPT_NAME=elf
-TEMPLATE_NAME=generic
-EXTRA_EM_FILE=genelf
+TEMPLATE_NAME=elf
 OUTPUT_FORMAT="elf32-fr30"
 NO_REL_RELOCS=yes
 TEXT_START_ADDR=0x10000
