commit f2240030da5201a0d93388640467801ead5810f0
Author: Jeff Law <jeffreyalaw@gmail.com>
Date:   Tue Nov 30 11:47:46 2021 -0700

    Regenerate

diff --git a/binutils/Makefile.in b/binutils/Makefile.in
index 336644a3983..f49da1afe17 100644
--- a/binutils/Makefile.in
+++ b/binutils/Makefile.in
@@ -2013,112 +2013,112 @@ install-exec-local: install-binPROGRAMS $(bin_PROGRAMS) $(noinst_PROGRAMS)
 	done
 
 # Man page generation from texinfo
-doc/addr2line.1:	$(binutils_TEXI)
+doc/addr2line.1:	$(binutils_TEXI) doc/$(am__dirstamp)
 	$(AM_V_GEN)touch $@
 	$(AM_V_at)-$(TEXI2POD) $(MANCONF) -Daddr2line < $(binutils_TEXI) > addr2line.pod
 	$(AM_V_at)-($(POD2MAN) addr2line.pod | sed -e '/^.if n .na/d' > $@.T$$$$ && \
 		mv -f $@.T$$$$ $@) || (rm -f $@.T$$$$ && exit 1)
 	$(AM_V_at)rm -f addr2line.pod
 
-doc/ar.1:	$(binutils_TEXI)
+doc/ar.1:	$(binutils_TEXI) doc/$(am__dirstamp)
 	$(AM_V_GEN)touch $@
 	$(AM_V_at)-$(TEXI2POD) $(MANCONF) -Dar < $(binutils_TEXI) > ar.pod
 	$(AM_V_at)-($(POD2MAN) ar.pod | sed -e '/^.if n .na/d' > $@.T$$$$ && \
 		mv -f $@.T$$$$ $@) || (rm -f $@.T$$$$ && exit 1)
 	$(AM_V_at)rm -f ar.pod
 
-doc/dlltool.1:	$(binutils_TEXI)
+doc/dlltool.1:	$(binutils_TEXI) doc/$(am__dirstamp)
 	$(AM_V_GEN)touch $@
 	$(AM_V_at)-$(TEXI2POD) $(MANCONF) -Ddlltool < $(binutils_TEXI) > dlltool.pod
 	$(AM_V_at)-($(POD2MAN) dlltool.pod | sed -e '/^.if n .na/d' > $@.T$$$$ && \
 		mv -f $@.T$$$$ $@) || (rm -f $@.T$$$$ && exit 1)
 	$(AM_V_at)rm -f dlltool.pod
 
-doc/nm.1:	$(binutils_TEXI)
+doc/nm.1:	$(binutils_TEXI) doc/$(am__dirstamp)
 	$(AM_V_GEN)touch $@
 	$(AM_V_at)-$(TEXI2POD) $(MANCONF) -Dnm < $(binutils_TEXI) > nm.pod
 	$(AM_V_at)-($(POD2MAN) nm.pod | sed -e '/^.if n .na/d' > $@.T$$$$ && \
 		mv -f $@.T$$$$ $@) || (rm -f $@.T$$$$ && exit 1)
 	$(AM_V_at)rm -f nm.pod
 
-doc/objcopy.1:	$(binutils_TEXI)
+doc/objcopy.1:	$(binutils_TEXI) doc/$(am__dirstamp)
 	$(AM_V_GEN)touch $@
 	$(AM_V_at)-$(TEXI2POD) $(MANCONF) -Dobjcopy < $(binutils_TEXI) > objcopy.pod
 	$(AM_V_at)-($(POD2MAN) objcopy.pod | sed -e '/^.if n .na/d' > $@.T$$$$ && \
 		mv -f $@.T$$$$ $@) || (rm -f $@.T$$$$ && exit 1)
 	$(AM_V_at)rm -f objcopy.pod
 
-doc/objdump.1:	$(binutils_TEXI)
+doc/objdump.1:	$(binutils_TEXI) doc/$(am__dirstamp)
 	$(AM_V_GEN)touch $@
 	$(AM_V_at)-$(TEXI2POD) $(MANCONF) -Dobjdump < $(binutils_TEXI) > objdump.pod
 	$(AM_V_at)-($(POD2MAN) objdump.pod | sed -e '/^.if n .na/d' > $@.T$$$$ && \
 		mv -f $@.T$$$$ $@) || (rm -f $@.T$$$$ && exit 1)
 	$(AM_V_at)rm -f objdump.pod
 
-doc/ranlib.1:	$(binutils_TEXI)
+doc/ranlib.1:	$(binutils_TEXI) doc/$(am__dirstamp)
 	$(AM_V_GEN)touch $@
 	$(AM_V_at)-$(TEXI2POD) $(MANCONF) -Dranlib < $(binutils_TEXI) > ranlib.pod
 	$(AM_V_at)-($(POD2MAN) ranlib.pod | sed -e '/^.if n .na/d' > $@.T$$$$ && \
 		mv -f $@.T$$$$ $@) || (rm -f $@.T$$$$ && exit 1)
 	$(AM_V_at)rm -f ranlib.pod
 
-doc/readelf.1:	$(binutils_TEXI)
+doc/readelf.1:	$(binutils_TEXI) doc/$(am__dirstamp)
 	$(AM_V_GEN)touch $@
 	$(AM_V_at)-$(TEXI2POD) $(MANCONF) -Dreadelf < $(binutils_TEXI) > readelf.pod
 	$(AM_V_at)-($(POD2MAN) readelf.pod | sed -e '/^.if n .na/d' > $@.T$$$$ && \
 		mv -f $@.T$$$$ $@) || (rm -f $@.T$$$$ && exit 1)
 	$(AM_V_at)rm -f readelf.pod
 
-doc/size.1:	$(binutils_TEXI)
+doc/size.1:	$(binutils_TEXI) doc/$(am__dirstamp)
 	$(AM_V_GEN)touch $@
 	$(AM_V_at)-$(TEXI2POD) $(MANCONF) -Dsize < $(binutils_TEXI) > size.pod
 	$(AM_V_at)-($(POD2MAN) size.pod | sed -e '/^.if n .na/d' > $@.T$$$$ && \
 		mv -f $@.T$$$$ $@) || (rm -f $@.T$$$$ && exit 1)
 	$(AM_V_at)rm -f size.pod
 
-doc/strings.1:	$(binutils_TEXI)
+doc/strings.1:	$(binutils_TEXI) doc/$(am__dirstamp)
 	$(AM_V_GEN)touch $@
 	$(AM_V_at)-$(TEXI2POD) $(MANCONF) -Dstrings < $(binutils_TEXI) > strings.pod
 	$(AM_V_at)-($(POD2MAN) strings.pod | sed -e '/^.if n .na/d' > $@.T$$$$ && \
 		mv -f $@.T$$$$ $@) || (rm -f $@.T$$$$ && exit 1)
 	$(AM_V_at)rm -f strings.pod
 
-doc/strip.1:	$(binutils_TEXI)
+doc/strip.1:	$(binutils_TEXI) doc/$(am__dirstamp)
 	$(AM_V_GEN)touch $@
 	$(AM_V_at)-$(TEXI2POD) $(MANCONF) -Dstrip < $(binutils_TEXI) > strip.pod
 	$(AM_V_at)-($(POD2MAN) strip.pod | sed -e '/^.if n .na/d' > $@.T$$$$ && \
 		mv -f $@.T$$$$ $@) || (rm -f $@.T$$$$ && exit 1)
 	$(AM_V_at)rm -f strip.pod
 
-doc/elfedit.1:	$(binutils_TEXI)
+doc/elfedit.1:	$(binutils_TEXI) doc/$(am__dirstamp)
 	$(AM_V_GEN)touch $@
 	$(AM_V_at)-$(TEXI2POD) $(MANCONF) -Delfedit < $(binutils_TEXI) > elfedit.pod
 	$(AM_V_at)-($(POD2MAN) elfedit.pod | sed -e '/^.if n .na/d' > $@.T$$$$ && \
 		mv -f $@.T$$$$ $@) || (rm -f $@.T$$$$ && exit 1)
 	$(AM_V_at)rm -f elfedit.pod
 
-doc/windres.1:	$(binutils_TEXI)
+doc/windres.1:	$(binutils_TEXI) doc/$(am__dirstamp)
 	$(AM_V_GEN)touch $@
 	$(AM_V_at)-$(TEXI2POD) $(MANCONF) -Dwindres < $(binutils_TEXI) > windres.pod
 	$(AM_V_at)-($(POD2MAN) windres.pod | sed -e '/^.if n .na/d' > $@.T$$$$ && \
 		mv -f $@.T$$$$ $@) || (rm -f $@.T$$$$ && exit 1)
 	$(AM_V_at)rm -f windres.pod
 
-doc/windmc.1:	$(binutils_TEXI)
+doc/windmc.1:	$(binutils_TEXI) doc/$(am__dirstamp)
 	$(AM_V_GEN)touch $@
 	$(AM_V_at)-$(TEXI2POD) $(MANCONF) -Dwindmc < $(binutils_TEXI) > windmc.pod
 	$(AM_V_at)-($(POD2MAN) windmc.pod | sed -e '/^.if n .na/d' > $@.T$$$$ && \
 		mv -f $@.T$$$$ $@) || (rm -f $@.T$$$$ && exit 1)
 	$(AM_V_at)rm -f windmc.pod
 
-doc/cxxfilt.man:	$(binutils_TEXI)
+doc/cxxfilt.man:	$(binutils_TEXI) doc/$(am__dirstamp)
 	$(AM_V_GEN)touch $@
 	$(AM_V_at)-$(TEXI2POD) $(MANCONF) -Dcxxfilt < $(binutils_TEXI) > $(DEMANGLER_NAME).pod
 	$(AM_V_at)-($(POD2MAN) $(DEMANGLER_NAME).pod | sed -e '/^.if n .na/d' > $@.T$$$$ && \
 		mv -f $@.T$$$$ $@) || (rm -f $@.T$$$$ && exit 1)
 	$(AM_V_at)rm -f $(DEMANGLER_NAME).pod
 
-doc/$(DEMANGLER_NAME).1: doc/cxxfilt.man Makefile
+doc/$(DEMANGLER_NAME).1: doc/cxxfilt.man Makefile doc/$(am__dirstamp)
 	$(AM_V_GEN)if test -f doc/cxxfilt.man; then \
 	  man=doc/cxxfilt.man; \
 	else \
